<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Presentation 2: Summary Statistics and Data Wrangling – R for Data Science</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">R for Data Science</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Course Material</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../about_heads.html"> 
<span class="menu-text">About us</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../presentations/presentation1.html">Presentations</a></li><li class="breadcrumb-item"><a href="../presentations/presentation2.html">Presentation 2: Summary Statistics and Data Wrangling</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../slides/R4datascience_slides.pdf" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Slides️</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Presentations</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../presentations/presentation1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Presentation 1: Data Cleaning</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../presentations/presentation2.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Presentation 2: Summary Statistics and Data Wrangling</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../presentations/presentation3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Presentation 3: Exploratory Data Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../presentations/presentation4A.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Presentation 4A: Conditions and For-loops</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../presentations/presentation4B.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Presentation 4B: Functions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../presentations/presentation5A.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Presentation 5A: Intro to Regression in R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../presentations/presentation5B.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Presentation 5B: Models and Model Evaluation in R</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false">
 <span class="menu-text">Exercises</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exercises/exercise0.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise 0: R script and Quarto</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exercises/exercise1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise 1: Data Cleaning</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exercises/exercise2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise 2: Summary Statistics and Data Wrangling</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exercises/exercise3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise 3: Exploratory Data Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exercises/exercise4A.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise 4A: Conditions and For-loops</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exercises/exercise4B.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise 4B: Functions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exercises/exercise5A.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise 5A: Intro to Regression in R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exercises/exercise5B.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise 5B: Models and Model Evaluation in R</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="false">
 <span class="menu-text">Solutions</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../tdhh/tdhh1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Solution 1: Data Cleaning</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../tdhh/tdhh2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Solution 2: Summary Statistics and Data Wrangling</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false">
 <span class="menu-text">Cheat Sheets</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../CheatSheet/quartoCheatSheet.pdf" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../CheatSheet/BaseRCheatSheet.pdf" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Base R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../CheatSheet/TidyverseCheatSheet.pdf" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Tidyverse</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../CheatSheet/ggplotCheatSheet.pdf" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ggplot2</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#load-packages-and-data" id="toc-load-packages-and-data" class="nav-link active" data-scroll-target="#load-packages-and-data">Load Packages and Data</a></li>
  <li><a href="#data-overview-and-ggplot2-recap" id="toc-data-overview-and-ggplot2-recap" class="nav-link" data-scroll-target="#data-overview-and-ggplot2-recap">Data Overview and <code>ggplot2</code> Recap</a></li>
  <li><a href="#formats-long-and-wide" id="toc-formats-long-and-wide" class="nav-link" data-scroll-target="#formats-long-and-wide">Formats: long and wide</a>
  <ul class="collapse">
  <li><a href="#the-long-format-is-ggplots-best-friend" id="toc-the-long-format-is-ggplots-best-friend" class="nav-link" data-scroll-target="#the-long-format-is-ggplots-best-friend">The Long Format is ggplot’s Best Friend</a></li>
  <li><a href="#pivot-back-into-wide-format" id="toc-pivot-back-into-wide-format" class="nav-link" data-scroll-target="#pivot-back-into-wide-format">Pivot back into wide format</a></li>
  </ul></li>
  <li><a href="#exploring-factors-categorical-variables" id="toc-exploring-factors-categorical-variables" class="nav-link" data-scroll-target="#exploring-factors-categorical-variables">Exploring Factors (Categorical Variables)</a></li>
  <li><a href="#summary-statistics" id="toc-summary-statistics" class="nav-link" data-scroll-target="#summary-statistics">Summary Statistics</a>
  <ul class="collapse">
  <li><a href="#using-across-and-everything-to-select-columns" id="toc-using-across-and-everything-to-select-columns" class="nav-link" data-scroll-target="#using-across-and-everything-to-select-columns">Using <code>across()</code> and <code>everything()</code> to select columns</a></li>
  <li><a href="#using-where-and-starts_with-to-select-numeric-columns" id="toc-using-where-and-starts_with-to-select-numeric-columns" class="nav-link" data-scroll-target="#using-where-and-starts_with-to-select-numeric-columns">Using <code>where()</code> and <code>starts_with()</code> to select numeric columns</a></li>
  <li><a href="#summarise-becomes-more-powerful" id="toc-summarise-becomes-more-powerful" class="nav-link" data-scroll-target="#summarise-becomes-more-powerful"><code>summarise()</code> becomes more powerful!</a></li>
  <li><a href="#the-anonymous-function-and-." id="toc-the-anonymous-function-and-." class="nav-link" data-scroll-target="#the-anonymous-function-and-.">The anonymous function: <code>~</code> and <code>.</code></a></li>
  <li><a href="#multiple-summary-statistics-with-na.rm-true" id="toc-multiple-summary-statistics-with-na.rm-true" class="nav-link" data-scroll-target="#multiple-summary-statistics-with-na.rm-true">Multiple Summary Statistics with <code>na.rm = TRUE</code></a></li>
  </ul></li>
  <li><a href="#handling-outliers" id="toc-handling-outliers" class="nav-link" data-scroll-target="#handling-outliers">Handling Outliers</a></li>
  <li><a href="#sample-vs-variable-outliers" id="toc-sample-vs-variable-outliers" class="nav-link" data-scroll-target="#sample-vs-variable-outliers">Sample vs variable outliers</a></li>
  <li><a href="#handling-missing-data" id="toc-handling-missing-data" class="nav-link" data-scroll-target="#handling-missing-data">Handling Missing Data</a>
  <ul class="collapse">
  <li><a href="#in-summary" id="toc-in-summary" class="nav-link" data-scroll-target="#in-summary">In Summary</a></li>
  <li><a href="#key-takeaways" id="toc-key-takeaways" class="nav-link" data-scroll-target="#key-takeaways">Key Takeaways</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../presentations/presentation1.html">Presentations</a></li><li class="breadcrumb-item"><a href="../presentations/presentation2.html">Presentation 2: Summary Statistics and Data Wrangling</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Presentation 2: Summary Statistics and Data Wrangling</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Now that we’ve cleaned our data, it’s time to dig deeper into the actual contents of our dataset. Lets do some Exploratory Data Analysis (EDA).</p>
<section id="load-packages-and-data" class="level2">
<h2 class="anchored" data-anchor-id="load-packages-and-data">Load Packages and Data</h2>
<p>Let’s begin by loading the packages we’ll need for data wrangling and plotting:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(factoextra)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, bring in the&nbsp;cleaned dataset&nbsp;we prepared earlier:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../data/Ovarian_comb_clean.RData"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the structure and first few rows</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(df_comb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "tbl_df"     "tbl"        "data.frame"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 54
  unique_patient_ID sample_recode primarysite summarygrade tumorstage substage
  &lt;chr&gt;             &lt;chr&gt;         &lt;fct&gt;       &lt;fct&gt;        &lt;fct&gt;      &lt;chr&gt;   
1 TCGA-20-0987      Tumor         ov          high         3          c       
2 TCGA-23-1031      Tumor         ov          high         4          &lt;NA&gt;    
3 TCGA-24-0979      Tumor         ov          high         4          &lt;NA&gt;    
4 TCGA-23-1117      Tumor         ov          high         3          c       
5 TCGA-23-1021      Tumor         ov          high         4          &lt;NA&gt;    
# ℹ 48 more variables: grade &lt;fct&gt;, age_at_initial_path_diagn &lt;int&gt;,
#   days_to_tumor_recurrence &lt;int&gt;, recurrence_status &lt;fct&gt;,
#   days_to_death &lt;int&gt;, vital_status &lt;fct&gt;, percent_normal_cells &lt;int&gt;,
#   percent_stromal_cells &lt;int&gt;, percent_tumor_cells &lt;int&gt;, batch &lt;fct&gt;,
#   percent_not_cancer_cells &lt;int&gt;, stage &lt;chr&gt;, dominant_cell_type &lt;chr&gt;,
#   group &lt;chr&gt;, COL10A1 &lt;dbl&gt;, COL11A1 &lt;dbl&gt;, COL11A2 &lt;dbl&gt;, COL13A1 &lt;dbl&gt;,
#   COL14A1 &lt;dbl&gt;, COL15A1 &lt;dbl&gt;, COL16A1 &lt;dbl&gt;, COL17A1 &lt;dbl&gt;, …</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>percent_not_cancer_cells)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="data-overview-and-ggplot2-recap" class="level2">
<h2 class="anchored" data-anchor-id="data-overview-and-ggplot2-recap">Data Overview and <code>ggplot2</code> Recap</h2>
<p>Depending on how we want to investigate the variables, there are many ways in EDA toolkit to explore variables:</p>
<ul>
<li><p>Use&nbsp;histograms&nbsp;or&nbsp;boxplots&nbsp;for single variables</p></li>
<li><p>Use&nbsp;scatterplots&nbsp;or&nbsp;barplots&nbsp;to check relationships between variables</p></li>
</ul>
<p>Let’s revisit some of the variables. This is also a good chance to refresh your&nbsp;<strong>ggplot2</strong>&nbsp;skills. If you need a detailed refresher, refer to the <a href="https://center-for-health-data-science.github.io/FromExceltoR/Presentations/presentation3.html">From Excel to R: Presentation 3</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Distribution of tumor cell percentage</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_comb, </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> percent_tumor_cells)) <span class="sc">+</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">30</span>, <span class="at">fill=</span><span class="st">"#482878FF"</span>) <span class="sc">+</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Count with barplot</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_comb, </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> grade, </span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> grade)) <span class="sc">+</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>(<span class="at">na.value =</span> <span class="st">"gray90"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Tumor percentage by summary grade</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_comb, </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> grade, </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">y =</span> percent_tumor_cells, </span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> grade)) <span class="sc">+</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>(<span class="at">na.value =</span> <span class="st">"gray90"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 22 rows containing non-finite outside the scale range
(`stat_boxplot()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>These simple plots give us an initial overview of the distribution of some of the variables - on their own and stratified by groups.</p>
</section>
<section id="formats-long-and-wide" class="level2">
<h2 class="anchored" data-anchor-id="formats-long-and-wide">Formats: long and wide</h2>
<p>Now, let’s say we want to create a single plot and compare several <strong>COL gene expression levels</strong> across different categories.</p>
<p>To plot and analyze more effectively, we need to reshape the data into <strong>long format</strong>, where each row represents a single observation:</p>
<p>The data can be reformatted to long format using <code>pivot_longer()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>df_comb_long <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span> </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">starts_with</span>(<span class="st">'COL1'</span>),</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"gene"</span>,</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"value"</span>)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>df_comb_long <span class="sc">%&gt;%</span> <span class="fu">select</span>(unique_patient_ID, age_at_initial_path_diagn, gene, value) <span class="sc">%&gt;%</span> <span class="fu">head</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  unique_patient_ID age_at_initial_path_diagn gene    value
  &lt;chr&gt;                                 &lt;int&gt; &lt;chr&gt;   &lt;dbl&gt;
1 TCGA-20-0987                             61 COL10A1  3.66
2 TCGA-20-0987                             61 COL11A1  3.20
3 TCGA-20-0987                             61 COL11A2  3.65
4 TCGA-20-0987                             61 COL13A1  4.07
5 TCGA-20-0987                             61 COL14A1  4.05
6 TCGA-20-0987                             61 COL15A1  4.98</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># one line per gene per person</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(df_comb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 578</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(df_comb_long)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6936</code></pre>
</div>
</div>
<section id="the-long-format-is-ggplots-best-friend" class="level3">
<h3 class="anchored" data-anchor-id="the-long-format-is-ggplots-best-friend">The Long Format is ggplot’s Best Friend</h3>
<p>With the reshaped <code>df_comb_long</code>, we can now create one <strong>combined plot</strong> that shows distributions for <strong>all genes</strong> in <strong>a single</strong> <code>ggplot</code> call. More context: add color-stratification by <code>summarygrade</code> and compare distributions side-by-side:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">na.omit</span>(df_comb_long), </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> gene, <span class="at">y =</span> value, <span class="at">fill =</span> summarygrade)) <span class="sc">+</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span> </span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Want histograms for all genes?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_comb_long, </span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> value, <span class="at">fill =</span> gene)) <span class="sc">+</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(gene), <span class="at">nrow =</span> <span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span> </span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This plot gives us a histogram for each gene, all in one go.</p>
<ul>
<li><p>No need to write separate plots manually.</p></li>
<li><p>Much easier to compare variables side-by-side.</p></li>
</ul>
</section>
<section id="pivot-back-into-wide-format" class="level3">
<h3 class="anchored" data-anchor-id="pivot-back-into-wide-format">Pivot back into wide format</h3>
<p>The <code>pivot_wider</code> function is used to transform data to wide format.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>df_comb_wide <span class="ot">&lt;-</span> df_comb_long <span class="sc">%&gt;%</span> </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> gene, </span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">values_from =</span> value)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_comb_wide)</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_comb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="exploring-factors-categorical-variables" class="level2">
<h2 class="anchored" data-anchor-id="exploring-factors-categorical-variables">Exploring Factors (Categorical Variables)</h2>
<p>When working with factor variables, we often want to know:</p>
<ul>
<li><p>Which <strong>levels</strong> (categories) exist</p></li>
<li><p>Whether groups are <strong>balanced</strong></p></li>
<li><p>How <strong>missing</strong> <strong>data</strong> overlaps</p></li>
</ul>
<p>Let’s have a look at the <code>vital_status</code> variable. It looks fairly balanced.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(df_comb<span class="sc">$</span>vital_status, <span class="at">useNA =</span> <span class="st">"ifany"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
deceased   living     &lt;NA&gt; 
     290      270       18 </code></pre>
</div>
</div>
<p>Now, let’s see how the <code>vital_status</code> levels are distributed across the <code>tumorstage</code> levels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(df_comb<span class="sc">$</span>vital_status, df_comb<span class="sc">$</span>stage, <span class="at">useNA =</span> <span class="st">"ifany"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          
           1-b 1-c 2-b 2-c 3-b 3-c 4-NA NA-NA
  deceased   1   2   2   5  12 214   52     2
  living     2  11   2  18  12 191   32     2
  &lt;NA&gt;       0   0   0   0   0   7    0    11</code></pre>
</div>
</div>
<p>Some stage categories could be merged for more power and clearer groups:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Relabel factor levels</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>df_comb <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">stage =</span> <span class="fu">fct_recode</span>(stage,</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"1"</span> <span class="ot">=</span> <span class="st">"1-b"</span>,</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"1"</span> <span class="ot">=</span> <span class="st">"1-c"</span>,</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"2"</span> <span class="ot">=</span> <span class="st">"2-b"</span>,</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"2"</span> <span class="ot">=</span> <span class="st">"2-c"</span>,</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"4"</span> <span class="ot">=</span> <span class="st">"4-NA"</span>,</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>                            <span class="at">NULL =</span> <span class="st">"NA-NA"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s see how other categorical variables relate to&nbsp;<code>vital_status</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot faceted bar plots colored by vital_status</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span> </span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">where</span>(is.factor), vital_status) <span class="sc">%&gt;%</span> </span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>vital_status, <span class="at">names_to =</span> <span class="st">"variable"</span>, <span class="at">values_to =</span> <span class="st">"value"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> value, <span class="at">fill =</span> vital_status)) <span class="sc">+</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(variable), <span class="at">scales =</span> <span class="st">"free_x"</span>) <span class="sc">+</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>After reviewing our categorical variables, we observed the following:</p>
<ul>
<li><p><strong><code>Tumorstage</code>&nbsp;and&nbsp;<code>grade</code></strong>&nbsp;are unbalanced. With certain stages and grades (e.g., 1 or 4) having low sample counts. This limits our ability to make reliable comparisons across all detailed levels.</p></li>
<li><p><strong><code>Batch</code> is evenly represented</strong>, which reduces concerns about batch effects.</p></li>
<li><p><strong><code>Vital_status</code> is nicely balanced</strong> making it well-suited for comparisons and modeling.</p></li>
</ul>
<p>These insights help us design our downstream analysis in a statistically sound and interpretable way.</p>
</section>
<section id="summary-statistics" class="level2">
<h2 class="anchored" data-anchor-id="summary-statistics">Summary Statistics</h2>
<p>It’s important to make sure our variables are well-behaved and ready for downstream analysis and modeling.</p>
<p>For this, let’s say we want to compute the mean of several columns. A basic (but tedious) approach might look like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_COL10A1 =</span> <span class="fu">mean</span>(COL10A1),</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean_COL11A1 =</span> <span class="fu">mean</span>(COL11A1),</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean_COL13A1 =</span> <span class="fu">mean</span>(COL13A1),</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean_COL17A1 =</span> <span class="fu">mean</span>(COL17A1),</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean_age_at_diagn =</span> <span class="fu">mean</span>(age_at_initial_path_diagn))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 5
  mean_COL10A1 mean_COL11A1 mean_COL13A1 mean_COL17A1 mean_age_at_diagn
         &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;             &lt;dbl&gt;
1         5.20         6.38         5.34         3.10                NA</code></pre>
</div>
</div>
<p>This works, but we need to name every column we want to apply <code>summarise</code> to. It’s verbose and error-prone — especially if you have dozens of variables.</p>
<p>Instead of looking at individual variable, we’ll introduce some tidyverse helper functions: <code>across()</code>, <code>where()</code>and <code>starts_with()</code> - that make summarizing variables much more efficient and scalable — so you don’t have to write repetitive code for every column.</p>
<p>These helpers are useful when we want to apply a functions, i.e.&nbsp;<code>summarise()</code> or <code>mutate()</code> to several columns.</p>
<section id="using-across-and-everything-to-select-columns" class="level3">
<h3 class="anchored" data-anchor-id="using-across-and-everything-to-select-columns">Using <code>across()</code> and <code>everything()</code> to select columns</h3>
<p>Lets select the columns which we want to apply <code>summarise</code> <strong>across</strong> in a dynamic fashion:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="at">.cols =</span> <span class="fu">everything</span>(), <span class="co"># Columns to run fuction on </span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.fns =</span> mean)) <span class="sc">%&gt;%</span> <span class="co"># Function </span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="dv">15</span><span class="sc">:</span><span class="dv">25</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 11
  percent_tumor_cells batch stage dominant_cell_type group COL10A1 COL11A1
                &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;              &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
1                  NA    NA    NA                 NA    NA    5.20    6.38
# ℹ 4 more variables: COL11A2 &lt;dbl&gt;, COL13A1 &lt;dbl&gt;, COL14A1 &lt;dbl&gt;,
#   COL15A1 &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
<section id="using-where-and-starts_with-to-select-numeric-columns" class="level3">
<h3 class="anchored" data-anchor-id="using-where-and-starts_with-to-select-numeric-columns">Using <code>where()</code> and <code>starts_with()</code> to select numeric columns</h3>
<p>We will probably not want to calculate means on non-numeric columns, so let’s select only numeric columns. For that we need another helper caller <code>where()</code> that lets us select columns based on their properties, like data type.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="at">.cols =</span> <span class="fu">where</span>(<span class="at">fn =</span> is.numeric), </span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.fns =</span> mean)) <span class="sc">%&gt;%</span> </span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="dv">5</span><span class="sc">:</span><span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 11
  percent_stromal_cells percent_tumor_cells COL10A1 COL11A1 COL11A2 COL13A1
                  &lt;dbl&gt;               &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
1                    NA                  NA    5.20    6.38    3.55    5.34
# ℹ 5 more variables: COL14A1 &lt;dbl&gt;, COL15A1 &lt;dbl&gt;, COL16A1 &lt;dbl&gt;,
#   COL17A1 &lt;dbl&gt;, COL18A1 &lt;dbl&gt;</code></pre>
</div>
</div>
<p>We can use <code>starts_with()</code> to select only columns starting with ‘COL’:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Columns that start with "COL"</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="fu">starts_with</span>(<span class="st">'COL'</span>), mean))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 34
  COL10A1 COL11A1 COL11A2 COL13A1 COL14A1 COL15A1 COL16A1 COL17A1 COL18A1
    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
1    5.20    6.38    3.55    5.34    5.01    5.89    4.95    3.10    7.26
# ℹ 25 more variables: COL19A1 &lt;dbl&gt;, COL1A1 &lt;dbl&gt;, COL1A2 &lt;dbl&gt;,
#   COL21A1 &lt;dbl&gt;, COL2A1 &lt;dbl&gt;, COL3A1 &lt;dbl&gt;, COL4A1 &lt;dbl&gt;, COL4A2 &lt;dbl&gt;,
#   COL4A3 &lt;dbl&gt;, COL4A3BP &lt;dbl&gt;, COL4A4 &lt;dbl&gt;, COL4A5 &lt;dbl&gt;, COL4A6 &lt;dbl&gt;,
#   COL5A1 &lt;dbl&gt;, COL5A2 &lt;dbl&gt;, COL5A3 &lt;dbl&gt;, COL6A1 &lt;dbl&gt;, COL6A2 &lt;dbl&gt;,
#   COL6A3 &lt;dbl&gt;, COL7A1 &lt;dbl&gt;, COL8A1 &lt;dbl&gt;, COL8A2 &lt;dbl&gt;, COL9A1 &lt;dbl&gt;,
#   COL9A2 &lt;dbl&gt;, COL9A3 &lt;dbl&gt;</code></pre>
</div>
</div>
<p>These helpers make your code cleaner, more scalable, and easier to maintain. They can be used to select columns in tidyverse, also outside of <code>across()</code>.</p>
</section>
<section id="summarise-becomes-more-powerful" class="level3">
<h3 class="anchored" data-anchor-id="summarise-becomes-more-powerful"><code>summarise()</code> becomes more powerful!</h3>
<p>So far, we’ve only applied a single function. But why stop at the mean? What if you want <strong>multiple statistics</strong> like mean, SD, min, and max - all in one go?</p>
<p>With <code>across()</code>, you can pass a <strong>list of functions</strong>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="at">.cols =</span> <span class="fu">starts_with</span>(<span class="st">"COL"</span>), </span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.fns =</span> <span class="fu">list</span>(mean, sd, min, max)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 136
  COL10A1_1 COL10A1_2 COL10A1_3 COL10A1_4 COL11A1_1 COL11A1_2 COL11A1_3
      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1      5.20      1.78      2.79      10.3      6.38      2.62      2.76
# ℹ 129 more variables: COL11A1_4 &lt;dbl&gt;, COL11A2_1 &lt;dbl&gt;, COL11A2_2 &lt;dbl&gt;,
#   COL11A2_3 &lt;dbl&gt;, COL11A2_4 &lt;dbl&gt;, COL13A1_1 &lt;dbl&gt;, COL13A1_2 &lt;dbl&gt;,
#   COL13A1_3 &lt;dbl&gt;, COL13A1_4 &lt;dbl&gt;, COL14A1_1 &lt;dbl&gt;, COL14A1_2 &lt;dbl&gt;,
#   COL14A1_3 &lt;dbl&gt;, COL14A1_4 &lt;dbl&gt;, COL15A1_1 &lt;dbl&gt;, COL15A1_2 &lt;dbl&gt;,
#   COL15A1_3 &lt;dbl&gt;, COL15A1_4 &lt;dbl&gt;, COL16A1_1 &lt;dbl&gt;, COL16A1_2 &lt;dbl&gt;,
#   COL16A1_3 &lt;dbl&gt;, COL16A1_4 &lt;dbl&gt;, COL17A1_1 &lt;dbl&gt;, COL17A1_2 &lt;dbl&gt;,
#   COL17A1_3 &lt;dbl&gt;, COL17A1_4 &lt;dbl&gt;, COL18A1_1 &lt;dbl&gt;, COL18A1_2 &lt;dbl&gt;, …</code></pre>
</div>
</div>
<p>This gives you one wide row per column, with new columns like <code>COL10A1_1</code>, <code>COL11A1_2</code>, etc. A bit cryptic, right?</p>
<p>Let’s clean it up by naming the functions and columns:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>gene_summary <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="at">.cols =</span> <span class="fu">starts_with</span>(<span class="st">"COL"</span>), </span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.fns =</span> <span class="fu">list</span>(<span class="at">mean =</span> mean, </span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>                               <span class="at">sd =</span> sd, </span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>                               <span class="at">min =</span> min, </span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>                               <span class="at">max =</span> max),</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.names =</span> <span class="st">"{.col}-{.fn}"</span>))</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>gene_summary</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 136
  `COL10A1-mean` `COL10A1-sd` `COL10A1-min` `COL10A1-max` `COL11A1-mean`
           &lt;dbl&gt;        &lt;dbl&gt;         &lt;dbl&gt;         &lt;dbl&gt;          &lt;dbl&gt;
1           5.20         1.78          2.79          10.3           6.38
# ℹ 131 more variables: `COL11A1-sd` &lt;dbl&gt;, `COL11A1-min` &lt;dbl&gt;,
#   `COL11A1-max` &lt;dbl&gt;, `COL11A2-mean` &lt;dbl&gt;, `COL11A2-sd` &lt;dbl&gt;,
#   `COL11A2-min` &lt;dbl&gt;, `COL11A2-max` &lt;dbl&gt;, `COL13A1-mean` &lt;dbl&gt;,
#   `COL13A1-sd` &lt;dbl&gt;, `COL13A1-min` &lt;dbl&gt;, `COL13A1-max` &lt;dbl&gt;,
#   `COL14A1-mean` &lt;dbl&gt;, `COL14A1-sd` &lt;dbl&gt;, `COL14A1-min` &lt;dbl&gt;,
#   `COL14A1-max` &lt;dbl&gt;, `COL15A1-mean` &lt;dbl&gt;, `COL15A1-sd` &lt;dbl&gt;,
#   `COL15A1-min` &lt;dbl&gt;, `COL15A1-max` &lt;dbl&gt;, `COL16A1-mean` &lt;dbl&gt;, …</code></pre>
</div>
</div>
<p>Much better! Now the column names are readable and include both the variable and the statistic.</p>
<p>But still not your preferred format? You can probably pivot your way out of that!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>gene_summary <span class="sc">%&gt;%</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">everything</span>(), </span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">"gene"</span>, <span class="st">"statistic"</span>), </span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_sep =</span> <span class="st">"-"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> statistic, </span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">values_from =</span> value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 34 × 5
   gene     mean    sd   min   max
   &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 COL10A1  5.20 1.78   2.79 10.3 
 2 COL11A1  6.38 2.62   2.76 11.4 
 3 COL11A2  3.55 0.260  3.09  5.41
 4 COL13A1  5.34 1.38   2.57 10.3 
 5 COL14A1  5.01 1.06   3.28  9.53
 6 COL15A1  5.89 1.02   3.42  8.68
 7 COL16A1  4.95 0.931  3.21  8.03
 8 COL17A1  3.10 0.214  2.71  5.57
 9 COL18A1  7.26 0.896  2.58 10.2 
10 COL19A1  2.75 0.150  2.40  3.31
# ℹ 24 more rows</code></pre>
</div>
</div>
<p>Now you get a long format table with one row per variable and all your stats in columns — clean, tidy, and ready for interpretation.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../figures/satisfying_meme.jpg" class="img-fluid quarto-figure quarto-figure-center figure-img" width="200"></p>
</figure>
</div>
</section>
<section id="the-anonymous-function-and-." class="level3">
<h3 class="anchored" data-anchor-id="the-anonymous-function-and-.">The anonymous function: <code>~</code> and <code>.</code></h3>
<p>We promised to get back to handling the <code>NA</code>s when doing summary stats. Let’s just add the <code>na.rm=TRUE</code> argument. To not have too many things going on at once we’ll only do <code>mean()</code> for now:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="at">.cols =</span> <span class="fu">where</span>(is.numeric), </span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.fns =</span> <span class="fu">list</span>(<span class="at">mean =</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)),</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.names =</span> <span class="st">"{.col}-{.fn}"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in `summarise()`:
ℹ In argument: `across(...)`.
Caused by error in `mean.default()`:
! argument "x" is missing, with no default</code></pre>
</div>
</div>
<p>Brrrtt! We may not.</p>
<p>Why doesn’t this work?</p>
<p>When you pass functions directly into <code>across()</code> using the <strong>shorthand syntax</strong> (<code>mean</code>, <code>sd</code>, etc.), you’re only allowed to use the <strong>bare function with default arguments</strong>. You will also notice that we didn’t use brackets after their names, which is part of using the function shorthanded. Once you try to add something like <code>na.rm = TRUE</code>, the shorthandness breaks.</p>
<p>To pass arguments to a function that is called inside another function (i.e.&nbsp;calling <code>mean</code> inside <code>summarise</code>), we need to use what’s called an <strong>anonymous function</strong>. Don’t worry — it’s not as scary as it sounds.</p>
<p>It is written as a <code>~</code> and looks like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="at">.cols =</span> <span class="fu">where</span>(is.numeric), </span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.fns =</span> <span class="fu">list</span>(<span class="at">mean =</span> <span class="sc">~</span> <span class="fu">mean</span>(., <span class="at">na.rm =</span> <span class="cn">TRUE</span>)),</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.names =</span> <span class="st">"{.col}-{.fn}"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 40
  `age_at_initial_path_diagn-mean` days_to_tumor_recurren…¹ `days_to_death-mean`
                             &lt;dbl&gt;                    &lt;dbl&gt;                &lt;dbl&gt;
1                             59.7                     624.                1010.
# ℹ abbreviated name: ¹​`days_to_tumor_recurrence-mean`
# ℹ 37 more variables: `percent_normal_cells-mean` &lt;dbl&gt;,
#   `percent_stromal_cells-mean` &lt;dbl&gt;, `percent_tumor_cells-mean` &lt;dbl&gt;,
#   `COL10A1-mean` &lt;dbl&gt;, `COL11A1-mean` &lt;dbl&gt;, `COL11A2-mean` &lt;dbl&gt;,
#   `COL13A1-mean` &lt;dbl&gt;, `COL14A1-mean` &lt;dbl&gt;, `COL15A1-mean` &lt;dbl&gt;,
#   `COL16A1-mean` &lt;dbl&gt;, `COL17A1-mean` &lt;dbl&gt;, `COL18A1-mean` &lt;dbl&gt;,
#   `COL19A1-mean` &lt;dbl&gt;, `COL1A1-mean` &lt;dbl&gt;, `COL1A2-mean` &lt;dbl&gt;, …</code></pre>
</div>
</div>
<p>Let’s break it down:</p>
<ul>
<li><p><code>~</code> to define the function. <code>~ mean(., na.rm = TRUE)</code> tells R:<br>
<em>“for each column, compute the mean, ignoring NAs”</em></p></li>
<li><p><code>.</code> is a placeholder for the current column being operated or ‘the data previously referred to’. We need to use the <code>.</code> because <code>mean</code> when called as a proper function needs to have an argument (a vector of numbers) to work on.</p></li>
</ul>
</section>
<section id="multiple-summary-statistics-with-na.rm-true" class="level3">
<h3 class="anchored" data-anchor-id="multiple-summary-statistics-with-na.rm-true">Multiple Summary Statistics with <code>na.rm = TRUE</code></h3>
<p>Now let’s compute <strong>several</strong> stats per column, all with <code>na.rm = TRUE</code>. Here we chose just the integer columns:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>stats <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="at">.cols =</span> <span class="fu">where</span>(is.integer),</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.fns =</span> <span class="fu">list</span>(<span class="at">mean =</span> <span class="sc">~</span> <span class="fu">mean</span>(., <span class="at">na.rm =</span> <span class="cn">TRUE</span>),  </span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>                               <span class="at">sd =</span> <span class="sc">~</span> <span class="fu">sd</span>(., <span class="at">na.rm =</span> <span class="cn">TRUE</span>), </span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>                               <span class="at">min =</span> <span class="sc">~</span> <span class="fu">min</span>(., <span class="at">na.rm =</span> <span class="cn">TRUE</span>), </span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>                               <span class="at">max =</span> <span class="sc">~</span> <span class="fu">max</span>(., <span class="at">na.rm =</span> <span class="cn">TRUE</span>)),</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.names =</span> <span class="st">"{.col}-{.fn}"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add reformating</span></span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">everything</span>(), </span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">"variable"</span>, <span class="st">"statistic"</span>), </span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_sep =</span> <span class="st">"-"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">value =</span> <span class="fu">round</span>(value, <span class="dv">1</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> statistic, </span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a>              <span class="at">values_from =</span> value) </span>
<span id="cb44-15"><a href="#cb44-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-16"><a href="#cb44-16" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(stats)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  variable                    mean    sd   min   max
  &lt;chr&gt;                      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 age_at_initial_path_diagn   59.7  11.6    26    89
2 days_to_tumor_recurrence   624.  608.      8  5480
3 days_to_death             1010.  803.      8  5480
4 percent_normal_cells         2.4   6.7     0    55
5 percent_stromal_cells       12.8  11.9     0    70
6 percent_tumor_cells         80.6  15.4     0   100</code></pre>
</div>
</div>
<p>This produces a tidy table with one row per gene and columns for <code>mean</code>, <code>sd</code>, <code>min</code>, and <code>max</code>.</p>
<p>The helper functions can be used in other tidyverse operation such as <code>mutate</code> and <code>select</code>.</p>
</section>
</section>
<section id="handling-outliers" class="level2">
<h2 class="anchored" data-anchor-id="handling-outliers">Handling Outliers</h2>
<p>Now that we’ve explored summary statistics, it’s time to consider <strong>outliers</strong> — data points that differ markedly from the rest.</p>
<p>Let’s use our tidyverse skills to <strong>calculate some basic stats</strong> and <strong>visualize potential outliers</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>stats <span class="ot">&lt;-</span> stats <span class="sc">%&gt;%</span> </span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">thr_upper =</span> mean <span class="sc">+</span> <span class="dv">2</span> <span class="sc">*</span> sd,</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">thr_lower =</span> mean <span class="sc">-</span> <span class="dv">2</span> <span class="sc">*</span> sd) <span class="sc">%&gt;%</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(thr_lower, thr_upper), <span class="at">names_to =</span> <span class="st">"thr_type"</span>, <span class="at">values_to =</span> <span class="st">"threshold"</span>)</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(stats)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12 × 7
   variable                    mean    sd   min   max thr_type  threshold
   &lt;chr&gt;                      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;         &lt;dbl&gt;
 1 age_at_initial_path_diagn   59.7  11.6    26    89 thr_lower      36.5
 2 age_at_initial_path_diagn   59.7  11.6    26    89 thr_upper      82.9
 3 days_to_tumor_recurrence   624.  608.      8  5480 thr_lower    -593. 
 4 days_to_tumor_recurrence   624.  608.      8  5480 thr_upper    1840. 
 5 days_to_death             1010.  803.      8  5480 thr_lower    -596. 
 6 days_to_death             1010.  803.      8  5480 thr_upper    2615. 
 7 percent_normal_cells         2.4   6.7     0    55 thr_lower     -11  
 8 percent_normal_cells         2.4   6.7     0    55 thr_upper      15.8
 9 percent_stromal_cells       12.8  11.9     0    70 thr_lower     -11  
10 percent_stromal_cells       12.8  11.9     0    70 thr_upper      36.6
11 percent_tumor_cells         80.6  15.4     0   100 thr_lower      49.8
12 percent_tumor_cells         80.6  15.4     0   100 thr_upper     111. </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>df_comb_longer <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span> </span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">where</span>(is.integer),</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"variable"</span>,</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"value"</span>)</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_comb_longer, </span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> value)) <span class="sc">+</span></span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">30</span>, <span class="at">fill=</span><span class="st">"#482878FF"</span>) <span class="sc">+</span></span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">data =</span> stats, </span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a>             <span class="fu">aes</span>(<span class="at">xintercept =</span> threshold),</span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a>             <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(variable), <span class="at">ncol =</span> <span class="dv">3</span>, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span> </span>
<span id="cb48-13"><a href="#cb48-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Sometimes, visualizing values in relation to other variables can make potential outliers easier to spot. Below, you’ll notice that a patient with 0% stromal cells and 0% tumor cells likely represents an outlier:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Bivariate scatter plot colored by stromal cell percentage</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_comb, </span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> percent_tumor_cells, </span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">y =</span> percent_stromal_cells, </span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> percent_normal_cells)) <span class="sc">+</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>() <span class="sc">+</span></span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We probably want to remove patients where both <code>percent_tumor_cells</code> and <code>percent_stromal_cells</code> are 0%.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(percent_tumor_cells <span class="sc">==</span> <span class="dv">0</span> <span class="sc">&amp;</span> percent_stromal_cells <span class="sc">==</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 7 × 53
  unique_patient_ID sample_recode primarysite summarygrade tumorstage substage
  &lt;chr&gt;             &lt;chr&gt;         &lt;fct&gt;       &lt;fct&gt;        &lt;fct&gt;      &lt;chr&gt;   
1 TCGA-24-1103      Tumor         ov          high         3          c       
2 TCGA-23-1119      Tumor         ov          high         3          c       
3 TCGA-23-1107      Tumor         ov          high         4          &lt;NA&gt;    
4 TCGA-23-1120      Tumor         ov          high         3          c       
5 TCGA-23-1121      Tumor         ov          high         3          c       
6 TCGA-23-1123      Tumor         ov          high         3          c       
7 TCGA-23-1124      Tumor         ov          high         3          c       
# ℹ 47 more variables: grade &lt;fct&gt;, age_at_initial_path_diagn &lt;int&gt;,
#   days_to_tumor_recurrence &lt;int&gt;, recurrence_status &lt;fct&gt;,
#   days_to_death &lt;int&gt;, vital_status &lt;fct&gt;, percent_normal_cells &lt;int&gt;,
#   percent_stromal_cells &lt;int&gt;, percent_tumor_cells &lt;int&gt;, batch &lt;fct&gt;,
#   stage &lt;fct&gt;, dominant_cell_type &lt;chr&gt;, group &lt;chr&gt;, COL10A1 &lt;dbl&gt;,
#   COL11A1 &lt;dbl&gt;, COL11A2 &lt;dbl&gt;, COL13A1 &lt;dbl&gt;, COL14A1 &lt;dbl&gt;, COL15A1 &lt;dbl&gt;,
#   COL16A1 &lt;dbl&gt;, COL17A1 &lt;dbl&gt;, COL18A1 &lt;dbl&gt;, COL19A1 &lt;dbl&gt;, COL1A1 &lt;dbl&gt;, …</code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>    (percent_tumor_cells <span class="sc">!=</span> <span class="dv">0</span> <span class="sc">|</span> <span class="fu">is.na</span>(percent_tumor_cells) <span class="sc">&amp;</span> </span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>            percent_stromal_cells <span class="sc">!=</span> <span class="dv">0</span>) <span class="sc">|</span> <span class="fu">is.na</span>(percent_stromal_cells))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sample-vs-variable-outliers" class="level2">
<h2 class="anchored" data-anchor-id="sample-vs-variable-outliers">Sample vs variable outliers</h2>
<p>There are many ways to&nbsp;detect and handle outliers. Instead of checking each variable separately, we can also look at&nbsp;<strong>all samples across multiple variables</strong>&nbsp;to spot unusual patterns.</p>
<p>One useful approach for datasets with many numeric or integer variables is to create a&nbsp;<strong>dendrogram</strong>&nbsp;using&nbsp;<strong>hierarchical clustering</strong>. This helps reveal samples that&nbsp;<strong>stand out</strong>&nbsp;from the rest.</p>
<p>Let´s try it, first, we select all&nbsp;integer columns and drop rows with missing values (they don’t help clustering).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>df_int <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">where</span>(is.integer)) <span class="sc">%&gt;%</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>()</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>df_int <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 6
  age_at_initial_path_diagn days_to_tumor_recurrence days_to_death
                      &lt;int&gt;                    &lt;int&gt;         &lt;int&gt;
1                        61                      442           701
2                        60                      574           574
3                        53                      428          1264
4                        78                       61            61
5                        74                      870           789
6                        73                       68            84
# ℹ 3 more variables: percent_normal_cells &lt;int&gt;, percent_stromal_cells &lt;int&gt;,
#   percent_tumor_cells &lt;int&gt;</code></pre>
</div>
</div>
<p>Next, we&nbsp;<strong>scale the data</strong>&nbsp;(details in&nbsp;<em>Presentation 3</em>), calculate pairwise distances between all pairs of scaled variables, and run&nbsp;<strong>hierarchical clustering</strong>&nbsp;to build a&nbsp;<strong>dendrogram</strong>.</p>
<p>This tree-like plot helps reveal samples that cluster far from the rest — potential outliers.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Euclidean pairwise distances</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>df_int_dist <span class="ot">&lt;-</span> df_int <span class="sc">%&gt;%</span> </span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">everything</span>(), scale)) <span class="sc">%&gt;%</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist</span>(., <span class="at">method =</span> <span class="st">'euclidean'</span>)</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Hierarchical clustering with Ward's distance metric</span></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>hclust_int <span class="ot">&lt;-</span> <span class="fu">hclust</span>(df_int_dist, <span class="at">method =</span> <span class="st">'average'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p>If the plot is hard to see:</p>
<ul>
<li><p>Copy-paste the code into the console to open it in the&nbsp;<strong>Plots</strong>&nbsp;pane.</p></li>
<li><p>Click&nbsp;<em>Export → Copy to Clipboard</em><strong>…</strong>&nbsp;and enlarge it as needed.</p></li>
</ul>
<p>Or save it as a high-resolution file and inspect it outside RStudio.</p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the result of clustering</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="co"># png("dendrogram_plot.png", width = 2000, height = 500)</span></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="co"># plot(hclust_int, main = "Clustering based on scaled integer values", cex = 0.7)</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a><span class="co"># dev.off()</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_dend</span>(hclust_int, <span class="at">h =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>From the dendrogram, we can see a few samples that branch off far from the main cluster — these long&nbsp;<strong>branch lengths (Height)</strong>&nbsp;may indicate outliers.</p>
<p>Let’s&nbsp;identify samples&nbsp;with unusually large heights and inspect them:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set a high threshold (top 0.1%)</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="co">#threshold &lt;- quantile(hclust_int$height, 0.99)</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>threshold <span class="ot">&lt;-</span> <span class="dv">6</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Find their order in the clustering</span></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>ord_inx <span class="ot">&lt;-</span> hclust_int<span class="sc">$</span>order[hclust_int<span class="sc">$</span>height <span class="sc">&gt;</span> threshold]</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>ord_inx</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 180  97</code></pre>
</div>
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># See which samples they are</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>df_int <span class="sc">%&gt;%</span> <span class="fu">slice</span>(ord_inx)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 6
  age_at_initial_path_diagn days_to_tumor_recurrence days_to_death
                      &lt;int&gt;                    &lt;int&gt;         &lt;int&gt;
1                        56                      213          1354
2                        51                      192           192
# ℹ 3 more variables: percent_normal_cells &lt;int&gt;, percent_stromal_cells &lt;int&gt;,
#   percent_tumor_cells &lt;int&gt;</code></pre>
</div>
</div>
<p>Next, we’ll bring back our&nbsp;<strong>summary statistics table</strong>&nbsp;to see whether these measurements fall outside what’s typical for the dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>stats</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12 × 7
   variable                    mean    sd   min   max thr_type  threshold
   &lt;chr&gt;                      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;         &lt;dbl&gt;
 1 age_at_initial_path_diagn   59.7  11.6    26    89 thr_lower      36.5
 2 age_at_initial_path_diagn   59.7  11.6    26    89 thr_upper      82.9
 3 days_to_tumor_recurrence   624.  608.      8  5480 thr_lower    -593. 
 4 days_to_tumor_recurrence   624.  608.      8  5480 thr_upper    1840. 
 5 days_to_death             1010.  803.      8  5480 thr_lower    -596. 
 6 days_to_death             1010.  803.      8  5480 thr_upper    2615. 
 7 percent_normal_cells         2.4   6.7     0    55 thr_lower     -11  
 8 percent_normal_cells         2.4   6.7     0    55 thr_upper      15.8
 9 percent_stromal_cells       12.8  11.9     0    70 thr_lower     -11  
10 percent_stromal_cells       12.8  11.9     0    70 thr_upper      36.6
11 percent_tumor_cells         80.6  15.4     0   100 thr_lower      49.8
12 percent_tumor_cells         80.6  15.4     0   100 thr_upper     111. </code></pre>
</div>
</div>
<p>Now we can&nbsp;compare these flagged samples&nbsp;to our summary statistics to understand&nbsp;why they stand out&nbsp;- and decide if they should be removed.</p>
<p>For these samples:</p>
<ul>
<li><p>Age at diagnosis&nbsp;and&nbsp;days to death&nbsp;are within a normal range.</p></li>
<li><p>However,&nbsp;percent stromal cells&nbsp;and&nbsp;percent tumor cells&nbsp;are unusually low.</p></li>
<li><p>These samples have&nbsp;uncharacteristically high normal cell proportions, which is extreme for this dataset.</p></li>
</ul>
<p>Given that they show&nbsp;extreme values&nbsp;across several key variables, it’s reasonable to treat them as&nbsp;outliers&nbsp;and&nbsp;remove them&nbsp;to avoid bias in our analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span>(</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>      (percent_normal_cells <span class="sc">==</span> <span class="dv">35</span> <span class="sc">&amp;</span> days_to_death <span class="sc">==</span> <span class="dv">1354</span>) <span class="sc">|</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>      (percent_normal_cells <span class="sc">==</span> <span class="dv">30</span> <span class="sc">&amp;</span> days_to_death <span class="sc">==</span> <span class="dv">192</span>)</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|</span></span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">is.na</span>(percent_normal_cells) <span class="sc">|</span> <span class="fu">is.na</span>(days_to_death))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p><strong>Be cautious when removing data</strong>: Outlier removal should always be guided by domain knowledge and clear justification. Removing too much — or for the wrong reasons — can distort your analysis.</p>
</blockquote>
</section>
<section id="handling-missing-data" class="level2">
<h2 class="anchored" data-anchor-id="handling-missing-data">Handling Missing Data</h2>
<p>Missing data is common in most real-world datasets and can significantly affect the quality of our analysis. During EDA, it’s essential to <strong>identify</strong>, <strong>understand</strong>, and <strong>properly handle</strong> missing values to avoid biased or misleading conclusions.</p>
<p>Let’s inspect rows with the most missing values:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">na_count =</span> <span class="fu">rowSums</span>(<span class="fu">is.na</span>(.))) <span class="sc">%&gt;%</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(na_count)) <span class="sc">%&gt;%</span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_head</span>(<span class="at">n =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 54
   unique_patient_ID sample_recode primarysite summarygrade tumorstage substage
   &lt;chr&gt;             &lt;chr&gt;         &lt;fct&gt;       &lt;fct&gt;        &lt;fct&gt;      &lt;chr&gt;   
 1 TCGA-01-0630      Healthy       &lt;NA&gt;        &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;    
 2 TCGA-01-0631      Healthy       &lt;NA&gt;        &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;    
 3 TCGA-01-0633      Healthy       &lt;NA&gt;        &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;    
 4 TCGA-01-0636      Healthy       &lt;NA&gt;        &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;    
 5 TCGA-01-0637      Healthy       &lt;NA&gt;        &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;    
 6 TCGA-01-0628      Healthy       &lt;NA&gt;        &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;    
 7 TCGA-01-0639      Healthy       &lt;NA&gt;        &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;    
 8 TCGA-01-0642      Healthy       &lt;NA&gt;        &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;    
 9 TCGA-04-1353      Tumor         &lt;NA&gt;        &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;    
10 TCGA-36-2539      Tumor         &lt;NA&gt;        &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;    
# ℹ 48 more variables: grade &lt;fct&gt;, age_at_initial_path_diagn &lt;int&gt;,
#   days_to_tumor_recurrence &lt;int&gt;, recurrence_status &lt;fct&gt;,
#   days_to_death &lt;int&gt;, vital_status &lt;fct&gt;, percent_normal_cells &lt;int&gt;,
#   percent_stromal_cells &lt;int&gt;, percent_tumor_cells &lt;int&gt;, batch &lt;fct&gt;,
#   stage &lt;fct&gt;, dominant_cell_type &lt;chr&gt;, group &lt;chr&gt;, COL10A1 &lt;dbl&gt;,
#   COL11A1 &lt;dbl&gt;, COL11A2 &lt;dbl&gt;, COL13A1 &lt;dbl&gt;, COL14A1 &lt;dbl&gt;, COL15A1 &lt;dbl&gt;,
#   COL16A1 &lt;dbl&gt;, COL17A1 &lt;dbl&gt;, COL18A1 &lt;dbl&gt;, COL19A1 &lt;dbl&gt;, COL1A1 &lt;dbl&gt;, …</code></pre>
</div>
</div>
<p>Looking at our missing values, we see that&nbsp;most missingness comes from samples labeled as&nbsp;<code>Healthy</code>(<code>sample_recode</code>).</p>
<p>This makes sense: Healthy tissue lacks tumor-specific details like&nbsp;<strong>tumor stage</strong>&nbsp;or&nbsp;<strong>percent tumor cells</strong>. This is an example of&nbsp;MAR (Missing At Random)&nbsp;— the missingness is related to an observed category (healthy vs.&nbsp;tumor tissue).</p>
<p>Since our focus is on&nbsp;tumor samples, not healthy tissue (few samples), we can&nbsp;remove these samples&nbsp;to clean up our data and avoid noise.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span> <span class="fu">filter</span>(sample_recode <span class="sc">==</span> <span class="st">"Tumor"</span>)</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Let’s inspect rows with the most missing values again.</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">na_count =</span> <span class="fu">rowSums</span>(<span class="fu">is.na</span>(.))) <span class="sc">%&gt;%</span></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(na_count)) <span class="sc">%&gt;%</span></span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_head</span>(<span class="at">n =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 54
   unique_patient_ID sample_recode primarysite summarygrade tumorstage substage
   &lt;chr&gt;             &lt;chr&gt;         &lt;fct&gt;       &lt;fct&gt;        &lt;fct&gt;      &lt;chr&gt;   
 1 TCGA-04-1353      Tumor         &lt;NA&gt;        &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;    
 2 TCGA-36-2539      Tumor         &lt;NA&gt;        &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;    
 3 TCGA-42-2593      Tumor         ov          &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;    
 4 TCGA-04-1341      Tumor         ov          high         &lt;NA&gt;       &lt;NA&gt;    
 5 TCGA-57-1994      Tumor         ov          &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;    
 6 TCGA-24-1104      Tumor         ov          high         4          &lt;NA&gt;    
 7 TCGA-23-1118      Tumor         ov          high         3          c       
 8 TCGA-24-1105      Tumor         ov          high         3          c       
 9 TCGA-23-1113      Tumor         ov          high         4          &lt;NA&gt;    
10 TCGA-13-0755      Tumor         ov          high         4          &lt;NA&gt;    
# ℹ 48 more variables: grade &lt;fct&gt;, age_at_initial_path_diagn &lt;int&gt;,
#   days_to_tumor_recurrence &lt;int&gt;, recurrence_status &lt;fct&gt;,
#   days_to_death &lt;int&gt;, vital_status &lt;fct&gt;, percent_normal_cells &lt;int&gt;,
#   percent_stromal_cells &lt;int&gt;, percent_tumor_cells &lt;int&gt;, batch &lt;fct&gt;,
#   stage &lt;fct&gt;, dominant_cell_type &lt;chr&gt;, group &lt;chr&gt;, COL10A1 &lt;dbl&gt;,
#   COL11A1 &lt;dbl&gt;, COL11A2 &lt;dbl&gt;, COL13A1 &lt;dbl&gt;, COL14A1 &lt;dbl&gt;, COL15A1 &lt;dbl&gt;,
#   COL16A1 &lt;dbl&gt;, COL17A1 &lt;dbl&gt;, COL18A1 &lt;dbl&gt;, COL19A1 &lt;dbl&gt;, COL1A1 &lt;dbl&gt;, …</code></pre>
</div>
</div>
<p>There are two main approaches when deciding what to do:</p>
<ul>
<li><p><strong>Remove rows</strong>&nbsp;with missing values</p></li>
<li><p><strong>Impute values</strong></p></li>
</ul>
<p>In our case:</p>
<p>We’ll replace numeric missing values with the median of each variable. We’ll replace missing factor levels with the most common category.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">c</span>(</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>      percent_normal_cells,</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>      percent_stromal_cells,</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>      percent_tumor_cells,</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>      age_at_initial_path_diagn,</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>      days_to_death,</span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>      days_to_tumor_recurrence</span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a>    <span class="sc">~</span> <span class="fu">if_else</span>(<span class="fu">is.na</span>(.x), <span class="fu">as.integer</span>(<span class="fu">median</span>(.x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)), <span class="fu">as.integer</span>(.x))</span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb68-13"><a href="#cb68-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-14"><a href="#cb68-14" aria-hidden="true" tabindex="-1"></a>df_comb <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span></span>
<span id="cb68-15"><a href="#cb68-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(</span>
<span id="cb68-16"><a href="#cb68-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">c</span>(stage, grade),</span>
<span id="cb68-17"><a href="#cb68-17" aria-hidden="true" tabindex="-1"></a>    <span class="sc">~</span> <span class="fu">fct_explicit_na</span>(.x, <span class="at">na_level =</span> <span class="fu">names</span>(<span class="fu">sort</span>(<span class="fu">table</span>(.x), <span class="at">decreasing =</span> <span class="cn">TRUE</span>))[<span class="dv">1</span>])</span>
<span id="cb68-18"><a href="#cb68-18" aria-hidden="true" tabindex="-1"></a>  ))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: There was 1 warning in `mutate()`.
ℹ In argument: `across(...)`.
Caused by warning:
! `fct_explicit_na()` was deprecated in forcats 1.0.0.
ℹ Please use `fct_na_value_to_level()` instead.</code></pre>
</div>
</div>
<p>We’ll keep the remaining&nbsp;<code>NA</code>s as-is for now.</p>
<blockquote class="blockquote">
<p>However always consider the impact of missing data. Even after imputing, missing data can cause uncertainty and bias so understands the result with caution.</p>
</blockquote>
<p><br></p>
<section id="in-summary" class="level3">
<h3 class="anchored" data-anchor-id="in-summary">In Summary</h3>
<p>We conducted an initial&nbsp;<strong>exploratory data analysis (EDA)</strong>&nbsp;and, overall, the dataset looks clean and well-structured.</p>
<ul>
<li><p><strong>Reshaped data</strong>&nbsp;— switched between wide and long formats to easily plot multiple variables together.</p></li>
<li><p><strong>Checked factors</strong>&nbsp;— reviewed category balance, merged sparse groups, and visualized how factors relate.</p></li>
<li><p><strong>Generated and inspected summary statistics</strong>&nbsp;— used tidyverse helpers to calculate means, standard deviations, and more for many columns at once, confirming that values align with biological expectations.</p></li>
<li><p><strong>Detected and removed outliers</strong>&nbsp;— used clustering to identify and exclude extreme or inconsistent samples with unusually high normal cell content.</p></li>
<li><p><strong>Handled missing data thoughtfully</strong>&nbsp;— removed irrelevant samples and imputed remaining gaps where appropriate.</p></li>
</ul>
</section>
<section id="key-takeaways" class="level3">
<h3 class="anchored" data-anchor-id="key-takeaways">Key Takeaways</h3>
<ul>
<li><p>Long format is your&nbsp;best friend&nbsp;when plotting or modeling.</p></li>
<li><p>Use&nbsp;<code>across()</code>&nbsp;and helpers to write&nbsp;clean, scalable summaries.</p></li>
<li><p>Visual checks are as important as statistical checks.</p></li>
<li><p>EDA is not optional&nbsp;— it’s the foundation of trustworthy analysis.</p></li>
<li><p>Keep your workflow&nbsp;reproducible&nbsp;and&nbsp;tidy&nbsp;for future modeling stages.</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(df_comb, <span class="at">file=</span><span class="st">"../data/Ovarian_comb_prep_Col.RData"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>