<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Presentation 2: Summary Statistics and Data Wrangling – R for Data Science</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">R for Data Science</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Course Material</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../about_heads.html"> 
<span class="menu-text">About us</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../presentations/presentation1.html">Presentations</a></li><li class="breadcrumb-item"><a href="../presentations/presentation2.html">Presentation 2: Summary Statistics and Data Wrangling</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../slides/R4datascience_slides.pdf" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Slides️</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Presentations</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../presentations/presentation1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Presentation 1: Data Cleaning</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../presentations/presentation2.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Presentation 2: Summary Statistics and Data Wrangling</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../presentations/presentation3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Presentation 3: Exploratory Data Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../presentations/presentation4A.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Presentation 4A: Conditions and For-loops</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../presentations/presentation4B.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Presentation 4B: Functions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../presentations/presentation5A.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Presentation 5A: Intro to Regression in R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../presentations/presentation5B.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Presentation 5B: Models and Model Evaluation in R</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../data/data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data️</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false">
 <span class="menu-text">Exercises</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exercises/exercise0.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise 0: R script and Quarto</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exercises/exercise1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise 1: Data Cleaning</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exercises/exercise2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise 2: Summary Statistics and Data Wrangling</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exercises/exercise3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise 3: Exploratory Data Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exercises/exercise4A.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise 4A: Conditions and For-loops</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exercises/exercise4B.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise 4B: Functions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exercises/exercise5A.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise 5A: Intro to Regression in R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exercises/exercise5B.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise 5B: Models and Model Evaluation in R</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="false">
 <span class="menu-text">Cheat Sheets</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../CheatSheet/quartoCheatSheet.pdf" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../CheatSheet/BaseRCheatSheet.pdf" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Base R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../CheatSheet/TidyverseCheatSheet.pdf" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Tidyverse</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../CheatSheet/ggplotCheatSheet.pdf" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ggplot2</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#load-packages-and-data" id="toc-load-packages-and-data" class="nav-link active" data-scroll-target="#load-packages-and-data">Load Packages and Data</a></li>
  <li><a href="#data-overview-and-ggplot2-recap" id="toc-data-overview-and-ggplot2-recap" class="nav-link" data-scroll-target="#data-overview-and-ggplot2-recap">Data Overview and <code>ggplot2</code> Recap</a>
  <ul class="collapse">
  <li><a href="#pipe-into-ggplot" id="toc-pipe-into-ggplot" class="nav-link" data-scroll-target="#pipe-into-ggplot">Pipe into <code>ggplot</code></a></li>
  <li><a href="#plotting-several-dataframes" id="toc-plotting-several-dataframes" class="nav-link" data-scroll-target="#plotting-several-dataframes">Plotting several dataframes</a></li>
  <li><a href="#plots-are-objects" id="toc-plots-are-objects" class="nav-link" data-scroll-target="#plots-are-objects">Plots are objects</a></li>
  </ul></li>
  <li><a href="#formats-long-and-wide" id="toc-formats-long-and-wide" class="nav-link" data-scroll-target="#formats-long-and-wide">Formats: long and wide</a>
  <ul class="collapse">
  <li><a href="#the-long-format-is-ggplots-best-friend" id="toc-the-long-format-is-ggplots-best-friend" class="nav-link" data-scroll-target="#the-long-format-is-ggplots-best-friend">The Long Format is ggplot’s Best Friend</a></li>
  <li><a href="#pivot-back-into-wide-format" id="toc-pivot-back-into-wide-format" class="nav-link" data-scroll-target="#pivot-back-into-wide-format">Pivot back into wide format</a></li>
  </ul></li>
  <li><a href="#exploring-factors-categorical-variables" id="toc-exploring-factors-categorical-variables" class="nav-link" data-scroll-target="#exploring-factors-categorical-variables">Exploring Factors (Categorical Variables)</a>
  <ul class="collapse">
  <li><a href="#pairwise-relationships-between-categorical-variables" id="toc-pairwise-relationships-between-categorical-variables" class="nav-link" data-scroll-target="#pairwise-relationships-between-categorical-variables">Pairwise Relationships Between Categorical Variables</a></li>
  </ul></li>
  <li><a href="#summary-statistics" id="toc-summary-statistics" class="nav-link" data-scroll-target="#summary-statistics">Summary Statistics</a>
  <ul class="collapse">
  <li><a href="#tidyverse-helpers-across-and-where" id="toc-tidyverse-helpers-across-and-where" class="nav-link" data-scroll-target="#tidyverse-helpers-across-and-where">Tidyverse helpers: <code>across()</code> and <code>where()</code></a></li>
  <li><a href="#using-across-to-select-columns" id="toc-using-across-to-select-columns" class="nav-link" data-scroll-target="#using-across-to-select-columns">Using <code>across()</code> to select columns</a></li>
  <li><a href="#using-where-to-select-numeric-columns" id="toc-using-where-to-select-numeric-columns" class="nav-link" data-scroll-target="#using-where-to-select-numeric-columns">Using <code>where()</code> to Select Numeric Columns</a></li>
  <li><a href="#other-column-selection-helpers" id="toc-other-column-selection-helpers" class="nav-link" data-scroll-target="#other-column-selection-helpers">Other Column Selection Helpers</a></li>
  <li><a href="#summarise-becomes-more-powerful" id="toc-summarise-becomes-more-powerful" class="nav-link" data-scroll-target="#summarise-becomes-more-powerful"><code>summarise()</code> becomes more powerful!</a></li>
  <li><a href="#the-anonymous-function-and-." id="toc-the-anonymous-function-and-." class="nav-link" data-scroll-target="#the-anonymous-function-and-.">The anonymous function: <code>~</code> and <code>.</code></a></li>
  <li><a href="#multiple-summary-statistics-with-na.rm-true" id="toc-multiple-summary-statistics-with-na.rm-true" class="nav-link" data-scroll-target="#multiple-summary-statistics-with-na.rm-true">Multiple Summary Statistics with <code>na.rm = TRUE</code></a></li>
  </ul></li>
  <li><a href="#handling-outliers" id="toc-handling-outliers" class="nav-link" data-scroll-target="#handling-outliers">Handling Outliers</a></li>
  <li><a href="#sample-outliers-across-variables" id="toc-sample-outliers-across-variables" class="nav-link" data-scroll-target="#sample-outliers-across-variables">Sample outliers across variables</a></li>
  <li><a href="#handling-missing-data" id="toc-handling-missing-data" class="nav-link" data-scroll-target="#handling-missing-data">Handling Missing Data</a></li>
  <li><a href="#in-summary" id="toc-in-summary" class="nav-link" data-scroll-target="#in-summary">In Summary</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../presentations/presentation1.html">Presentations</a></li><li class="breadcrumb-item"><a href="../presentations/presentation2.html">Presentation 2: Summary Statistics and Data Wrangling</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Presentation 2: Summary Statistics and Data Wrangling</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Now that we’ve cleaned our data, it’s time to dig deeper into the actual contents of our dataset. Exploratory Data Analysis (EDA) is an essential step in any data analysis or modeling workflow.</p>
<p>EDA is about getting an overview of your data before fitting it to any models. It helps us:</p>
<ul>
<li><p>Understand <strong>distributions</strong></p></li>
<li><p>Spot <strong>outliers</strong></p></li>
<li><p>Reveal <strong>patterns</strong></p></li>
<li><p>Explore <strong>relationships</strong> between variables</p></li>
</ul>
<p>There are three broad ways to investigate variables:</p>
<ol type="1">
<li><p><strong>Univariate</strong> – Examining one variable at a time.</p></li>
<li><p><strong>Bivariate</strong> – Exploring relationships between two variables.</p></li>
<li><p><strong>Multivariate</strong> – Looking at relationships among multiple variables.</p></li>
</ol>
<section id="load-packages-and-data" class="level2">
<h2 class="anchored" data-anchor-id="load-packages-and-data">Load Packages and Data</h2>
<p>Let’s begin by loading the packages we’ll need for data wrangling and plotting:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse) </span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># library(ggforce) </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we load our dataset which was pre-processed and saved as an <code>.RData</code> file at the end of Presentation 1:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../data/Ovarian_comb_clean.RData"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Quick peek at the structure and contents</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(df_comb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "tbl_df"     "tbl"        "data.frame"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(df_comb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>tibble [578 × 29] (S3: tbl_df/tbl/data.frame)
 $ unique_patient_ID        : chr [1:578] "TCGA-20-0987" "TCGA-23-1031" "TCGA-24-0979" "TCGA-23-1117" ...
 $ sample_recode            : chr [1:578] "Tumor" "Tumor" "Tumor" "Tumor" ...
 $ summarygrade             : Factor w/ 2 levels "high","low": 1 1 1 1 1 2 1 1 1 1 ...
 $ summarystage             : Factor w/ 2 levels "early","late": 2 2 2 2 2 2 2 2 2 2 ...
 $ tumorstage               : Factor w/ 4 levels "1","2","3","4": 3 4 4 3 4 3 3 4 3 3 ...
 $ grade                    : Factor w/ 4 levels "1","2","3","4": 3 3 3 3 3 2 3 3 3 3 ...
 $ age_at_initial_path_diagn: int [1:578] 61 60 53 42 45 78 74 73 45 45 ...
 $ recurrence_status        : Factor w/ 2 levels "norecurrence",..: 2 1 2 2 1 1 2 1 1 2 ...
 $ days_to_death            : int [1:578] 701 574 1264 1013 1446 61 789 84 2616 816 ...
 $ vital_status             : Factor w/ 2 levels "deceased","living": 1 1 1 1 1 1 2 1 2 2 ...
 $ percent_normal_cells     : int [1:578] 16 8 0 0 0 0 12 15 NA 0 ...
 $ percent_stromal_cells    : int [1:578] 1 5 25 NA 0 36 2 5 NA NA ...
 $ percent_tumor_cells      : int [1:578] 72 80 72 NA NA 64 84 72 NA NA ...
 $ batch                    : Factor w/ 14 levels "9","11","12",..: 3 3 3 3 3 3 3 3 3 3 ...
 $ percent_not_cancer_cells : int [1:578] 17 13 25 NA 0 36 14 20 NA NA ...
 $ stage                    : chr [1:578] "3-high" "4-high" "4-high" "3-high" ...
 $ dominant_cell_type       : chr [1:578] "cancer_cells" "cancer_cells" "cancer_cells" NA ...
 $ group                    : chr [1:578] "Late stage - Mature" "Late stage - Mature" "Late stage - Young" "Late stage - Young" ...
 $ CXCL1                    : num [1:578] 4.97 3.99 3.87 3.16 3.35 ...
 $ CXCL10                   : num [1:578] 9.91 6.61 10 9.24 7.86 ...
 $ CXCL11                   : num [1:578] 5.26 2.9 4.19 3.89 2.95 ...
 $ CXCL12                   : num [1:578] 5.44 6.31 4.19 9.07 5.1 ...
 $ CXCL13                   : num [1:578] 6.55 3.18 3.36 5.1 3.41 ...
 $ CXCL14                   : num [1:578] 3.07 2.71 2.77 6.2 5.13 ...
 $ CXCL2                    : num [1:578] 5.14 5.12 7.28 5.88 5.5 ...
 $ CXCL3                    : num [1:578] 3.72 3.76 3.88 3.5 3.69 ...
 $ CXCL5                    : num [1:578] 3.31 3.11 3.38 3.04 2.97 ...
 $ CXCL6                    : num [1:578] 3.17 2.95 2.9 3.08 2.58 ...
 $ CXCL9                    : num [1:578] 9.19 3.49 4.83 8.82 3.27 ...</code></pre>
</div>
</div>
<p>Depending on how we want to investigate the variables, there are different tools in your EDA toolkit we can employ:</p>
<p>Univariate Analysis</p>
<ul>
<li><p><strong>Histograms</strong> – Check distributions of numeric variables</p></li>
<li><p><strong>Boxplots</strong> – Detect outliers and skewness</p></li>
<li><p><strong>Bar charts</strong> – Explore frequency of categorical variables</p></li>
<li><p><strong>Summary stats</strong> – Mean, median, standard deviation (SD), min, max, interquartile range (IQR)</p></li>
</ul>
<p>Bivariate Analysis</p>
<ul>
<li><p><strong>Scatter plots</strong> – Explore relationships between two numeric variables</p></li>
<li><p><strong>Correlation coefficients</strong> – Assess linear strength (e.g., Pearson’s <em>r</em>)</p></li>
<li><p><strong>Contingency tables</strong> – Analyze relationships between categorical variables</p></li>
<li><p><strong>Line graphs</strong> – Track trends over time</p></li>
</ul>
</section>
<section id="data-overview-and-ggplot2-recap" class="level2">
<h2 class="anchored" data-anchor-id="data-overview-and-ggplot2-recap">Data Overview and <code>ggplot2</code> Recap</h2>
<p>Let’s revisit some of the variables in the dataset and refresh our <code>ggplot2</code> skills.</p>
<p>If you need a detailed refresher, refer to the <a href="https://center-for-health-data-science.github.io/FromExceltoR/Presentations/presentation3.html">From Excel to R: Presentation 3</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Distribution of tumor cell percentage</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_comb, </span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> percent_tumor_cells)) <span class="sc">+</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Tumor percentage by summary grade</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_comb, </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> grade, </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">y =</span> percent_tumor_cells, </span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> grade)) <span class="sc">+</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>(<span class="at">na.value =</span> <span class="st">"grey50"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-3-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Count with barplot</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_comb, </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> grade, </span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> grade)) <span class="sc">+</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>(<span class="at">na.value =</span> <span class="st">"grey50"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-3-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>These simple plots give us an initial overview of the distribution of some of the variables - on their own and stratified by groups.</p>
<section id="pipe-into-ggplot" class="level3">
<h3 class="anchored" data-anchor-id="pipe-into-ggplot">Pipe into <code>ggplot</code></h3>
<p>Tired of cluttering your environment with dozens of intermediate data frames? There’s a solution: you can pipe (<code>%&gt;%</code>) your filtered data directly into <code>ggplot()</code> without saving it first. When you do this, you don’t need to specify the <code>data = ...</code> argument inside <code>ggplot()</code> — the data is being <em>piped</em> into the function automatically.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(sample_recode  <span class="sc">==</span> <span class="st">"Tumor"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> CXCL11, </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> CXCL10, </span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">color =</span> summarygrade)) <span class="sc">+</span> </span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>(<span class="at">na.value =</span> <span class="st">"grey50"</span>) <span class="sc">+</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="plotting-several-dataframes" class="level3">
<h3 class="anchored" data-anchor-id="plotting-several-dataframes">Plotting several dataframes</h3>
<p>Sometimes we want to add extra layers of information to a plot — for example, combining raw data distributions with summary statistics.</p>
<p>In this case, we’ll create a violin plot to show the distribution of the <code>age_at_initial_path_diagn</code> values stratified by the <code>summarygrade</code> categories. At the same time, we want to overlay the mean <code>age_at_initial_path_diagn</code> value for each <code>summarygrade</code> category as a red point.</p>
<p>We start by summarizing our data to get the mean <code>age_at_initial_path_diagn</code> for each <code>summarygrade</code> category:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>gene_mean <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(grade) <span class="sc">%&gt;%</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean =</span> <span class="fu">mean</span>(age_at_initial_path_diagn)) <span class="sc">%&gt;%</span> </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(grade))</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>gene_mean</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 2
  grade  mean
  &lt;fct&gt; &lt;dbl&gt;
1 1      50.2
2 2      58.4
3 3      60.0
4 4      51  </code></pre>
</div>
</div>
<p>Now we plot the distributions with <code>geom_violin()</code> and add the mean values using a second data frame in <code>geom_point()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span> </span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span>(<span class="fu">is.na</span>(age_at_initial_path_diagn) <span class="sc">|</span> <span class="fu">is.na</span>(grade))) <span class="sc">%&gt;%</span> </span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> grade, </span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> age_at_initial_path_diagn, </span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">fill =</span> grade)) <span class="sc">+</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>() <span class="sc">+</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> gene_mean, <span class="fu">aes</span>(<span class="at">x =</span> grade, <span class="at">y =</span> mean), <span class="at">color =</span> <span class="st">'red'</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">'Age at Initial Pathologic Diagnosis by grade'</span>, </span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">caption =</span> <span class="st">'Only one individual in grade 4.'</span>) <span class="sc">+</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><code>geom_point(data = gene_mean)</code> tells <code>ggplot</code> to use a <strong>different data frame</strong> for this layer — a powerful way to combine data sources in a single plot.</p>
<p>This makes it easy to visualize both <strong>variation</strong> and <strong>central tendency</strong> at the same time.</p>
</section>
<section id="plots-are-objects" class="level3">
<h3 class="anchored" data-anchor-id="plots-are-objects">Plots are objects</h3>
<p>Just like data frames, <code>ggplot</code> plots are R objects. You can assign them to variables and display them later:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>awesome_plot <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span> </span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span>(<span class="fu">is.na</span>(age_at_initial_path_diagn) <span class="sc">|</span> <span class="fu">is.na</span>(grade))) <span class="sc">%&gt;%</span> </span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> grade,</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> age_at_initial_path_diagn,</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">fill =</span> grade)) <span class="sc">+</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>() <span class="sc">+</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> gene_mean, <span class="fu">aes</span>(<span class="at">x =</span> grade, <span class="at">y=</span>mean), <span class="at">color =</span> <span class="st">'red'</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span> </span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">'Age at Initial Pathologic Diagnosis by grade'</span>, </span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">caption =</span> <span class="st">'Only one individual in grade 4.'</span>) <span class="sc">+</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If R is being every pesky about showing you plots (e.g.&nbsp;if you want to display them in a loop) wrapping <code>print()</code> around the name of the plot object usually helps:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(awesome_plot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This is especially useful for creating multiple plots programmatically.</p>
</section>
</section>
<section id="formats-long-and-wide" class="level2">
<h2 class="anchored" data-anchor-id="formats-long-and-wide">Formats: long and wide</h2>
<p>The doctrine of <code>ggplot</code> is that every information you want to plot must be in a column. There is one column that describes the x-axis, one for the y-axis, and one for each additional aesthetic like color, size, shape, ect.</p>
<p>Now, let’s say we want to create a single plot and compare several <strong>CXCL gene expression levels</strong> across different categories.</p>
<p>In the current <strong>wide format</strong> (like most spreadsheets), each gene is stored in its own column:</p>
<table class="caption-top">
<thead>
<tr class="header">
<th>ID</th>
<th>CXCL1</th>
<th>CXCL2</th>
<th>CXCL3</th>
<th>…</th>
</tr>
</thead>
<tbody>
</tbody>
</table>
<p>This layout is fine for browsing spreadsheets, but to create the plot described above you’d have to write a separate plot for each gene/column — not efficient.</p>
<p>To plot and analyze more effectively, we need to reshape the data into <strong>long format</strong>, where each row represents a single observation:</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>ID</th>
<th>gene</th>
<th>value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>CXCL1</td>
<td>2.3</td>
</tr>
<tr class="even">
<td>1</td>
<td>CXCL2</td>
<td>1.9</td>
</tr>
<tr class="odd">
<td>2</td>
<td>CXCL1</td>
<td>3.1</td>
</tr>
<tr class="even">
<td>…</td>
<td>…</td>
<td>…</td>
</tr>
</tbody>
</table>
<p>We can do this using <code>pivot_longer()</code>:</p>
<p>The data can be reformatted to long format such that there is one line per gene per person. For this we use the <code>pivot_longer</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>df_comb_long <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span> </span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">starts_with</span>(<span class="st">'CXCL'</span>),</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"gene"</span>,</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"value"</span>)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>df_comb_long <span class="sc">%&gt;%</span> <span class="fu">select</span>(unique_patient_ID, age_at_initial_path_diagn, gene, value) <span class="sc">%&gt;%</span> <span class="fu">head</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  unique_patient_ID age_at_initial_path_diagn gene   value
  &lt;chr&gt;                                 &lt;int&gt; &lt;chr&gt;  &lt;dbl&gt;
1 TCGA-20-0987                             61 CXCL1   4.97
2 TCGA-20-0987                             61 CXCL10  9.91
3 TCGA-20-0987                             61 CXCL11  5.26
4 TCGA-20-0987                             61 CXCL12  5.44
5 TCGA-20-0987                             61 CXCL13  6.55
6 TCGA-20-0987                             61 CXCL14  3.07</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># one line per gene per person</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(df_comb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 578</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(df_comb_long)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6358</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>df_comb_long<span class="sc">$</span>gene <span class="sc">%&gt;%</span> <span class="fu">unique</span>() <span class="sc">%&gt;%</span> <span class="fu">length</span>() <span class="sc">*</span> <span class="fu">nrow</span>(df_comb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6358</code></pre>
</div>
</div>
<section id="the-long-format-is-ggplots-best-friend" class="level3">
<h3 class="anchored" data-anchor-id="the-long-format-is-ggplots-best-friend">The Long Format is ggplot’s Best Friend</h3>
<p>With the reshaped <code>df_comb_long</code>, we can now create one <strong>combined plot</strong> that shows distributions for <strong>all genes</strong> in <strong>a single</strong> <code>ggplot</code> call:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_comb_long, </span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> gene, <span class="at">y =</span> value, <span class="at">fill =</span> gene)) <span class="sc">+</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span> </span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Want histograms for all genes?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_comb_long, </span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> value)) <span class="sc">+</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(gene), <span class="at">nrow =</span> <span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This plot gives us a histogram for each gene, all in one go.</p>
<ul>
<li><p>No need to write separate plots manually.</p></li>
<li><p>Much easier to compare variables side-by-side.</p></li>
</ul>
<p>More context: add color-stratification by <code>summarygrade</code> and compare distributions side-by-side:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">na.omit</span>(df_comb_long), </span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> gene, <span class="at">y =</span> value, <span class="at">fill =</span> summarygrade)) <span class="sc">+</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span> </span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span> </span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="pivot-back-into-wide-format" class="level3">
<h3 class="anchored" data-anchor-id="pivot-back-into-wide-format">Pivot back into wide format</h3>
<p>The <code>pivot_wider</code> function is used to transform data to wide format.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>df_comb_wide <span class="ot">&lt;-</span> df_comb_long <span class="sc">%&gt;%</span> </span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> gene, </span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">values_from =</span> value)</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_comb_wide)</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_comb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="exploring-factors-categorical-variables" class="level2">
<h2 class="anchored" data-anchor-id="exploring-factors-categorical-variables">Exploring Factors (Categorical Variables)</h2>
<p>When working with <strong>factor variables</strong>, we often want to know:</p>
<ul>
<li><p>Which levels (categories) exist</p></li>
<li><p>Whether groups are <strong>balanced</strong></p></li>
<li><p>How <strong>missingness</strong> overlaps across factors</p></li>
</ul>
<p>Let’s have a look at the <code>vital_status</code> variable. It looks fairly balanced.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(df_comb<span class="sc">$</span>vital_status, <span class="at">useNA =</span> <span class="st">"ifany"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
deceased   living     &lt;NA&gt; 
     290      270       18 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span> </span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(vital_status)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  vital_status     n
  &lt;fct&gt;        &lt;int&gt;
1 deceased       290
2 living         270
3 &lt;NA&gt;            18</code></pre>
</div>
</div>
<p>Now, let’s see how the <code>vital_status</code> levels are distributed across the <code>tumorstage</code> levels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(df_comb<span class="sc">$</span>vital_status, df_comb<span class="sc">$</span>tumorstage, <span class="at">useNA =</span> <span class="st">"ifany"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          
             1   2   3   4 &lt;NA&gt;
  deceased   3   7 226  52    2
  living    13  20 203  32    2
  &lt;NA&gt;       0   0   7   0   11</code></pre>
</div>
</div>
<section id="pairwise-relationships-between-categorical-variables" class="level3">
<h3 class="anchored" data-anchor-id="pairwise-relationships-between-categorical-variables">Pairwise Relationships Between Categorical Variables</h3>
<p>We will now have a look at the distribution of levels of the factor variables across <code>vital_status</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Keep only factor variables including vital_status</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span> </span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">where</span>(is.factor), vital_status) <span class="sc">%&gt;%</span> </span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>vital_status, <span class="at">names_to =</span> <span class="st">"variable"</span>, <span class="at">values_to =</span> <span class="st">"value"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Plot faceted bar plots colored by vital_status</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> value, <span class="at">fill =</span> vital_status)) <span class="sc">+</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(variable), <span class="at">scales =</span> <span class="st">"free_x"</span>) <span class="sc">+</span></span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="after-reviewing-our-categorical-variables-we-observed-the-following" class="level4">
<h4 class="anchored" data-anchor-id="after-reviewing-our-categorical-variables-we-observed-the-following">After reviewing our categorical variables, we observed the following:</h4>
<ul>
<li><p><strong><code>Tumorstage</code> and <code>grade</code> are imbalanced</strong> — with certain stages and grades (e.g., 1 or 4) having low sample counts. This limits our ability to make reliable comparisons across all detailed levels.</p>
<ul>
<li><p>To address this, we may use broader categories:</p>
<ul>
<li><p><em>Early vs.&nbsp;Late Stage</em> (e.g., Stage 1–2 vs.&nbsp;3-4)</p></li>
<li><p><em>Low vs.&nbsp;High Grade</em> (e.g., Grade 1–2 vs.&nbsp;3–4)</p></li>
</ul></li>
</ul></li>
<li><p><strong><code>Batch</code> is evenly represented</strong>, which reduces concerns about batch effects in our analysis.</p></li>
<li><p><strong><code>Vital_status</code> is nicely balanced</strong> making it well-suited for comparisons and modeling.</p></li>
</ul>
<p>These insights will help us structure our downstream analysis in a statistically sound and interpretable way.</p>
</section>
</section>
</section>
<section id="summary-statistics" class="level2">
<h2 class="anchored" data-anchor-id="summary-statistics">Summary Statistics</h2>
<p>Even after initial data cleaning and harmonization, it’s important to make sure our variables are well-behaved and ready for downstream analysis and modeling.</p>
<p>Instead of looking at individual variable, we’ll introduce some tidyverse helper functions that make summarizing variables much more efficient and scalable — so you don’t have to write repetitive code for every column.</p>
<section id="tidyverse-helpers-across-and-where" class="level3">
<h3 class="anchored" data-anchor-id="tidyverse-helpers-across-and-where">Tidyverse helpers: <code>across()</code> and <code>where()</code></h3>
<p>In this section we will say a bit more about the so called tidyverse helpers such as <code>across()</code>, <code>where()</code>, and <code>starts_with()</code>. These helpers are useful when we want to apply a functions, i.e.&nbsp;<code>summarise()</code> or <code>mutate()</code> to several columns.</p>
</section>
<section id="using-across-to-select-columns" class="level3">
<h3 class="anchored" data-anchor-id="using-across-to-select-columns">Using <code>across()</code> to select columns</h3>
<p>Let’s say we want to compute the mean of several columns. A basic (but tedious) approach might look like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_CXCL1 =</span> <span class="fu">mean</span>(CXCL1),</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean_CXCL2 =</span> <span class="fu">mean</span>(CXCL2),</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean_CXCL3 =</span> <span class="fu">mean</span>(CXCL3),</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean_CXCL5 =</span> <span class="fu">mean</span>(CXCL5),</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean_age_at_diagn =</span> <span class="fu">mean</span>(age_at_initial_path_diagn))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 5
  mean_CXCL1 mean_CXCL2 mean_CXCL3 mean_CXCL5 mean_age_at_diagn
       &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;             &lt;dbl&gt;
1       4.72       5.88       3.82       3.43                NA</code></pre>
</div>
</div>
<p>This works, but we need to name every column we want to apply <code>summarise</code> to. It’s verbose and error-prone — especially if you have dozens of variables.</p>
<p>Enter <code>across()</code>: lets us select the columns which we want to apply <code>summarise</code> <strong>across</strong> in a dynamic fashion:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="at">.cols =</span> <span class="fu">everything</span>(), <span class="co"># Columns to run fuction on </span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.fns =</span> mean)) <span class="sc">%&gt;%</span> <span class="co"># Function </span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="dv">15</span><span class="sc">:</span><span class="dv">25</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 11
  percent_not_cancer_cells stage dominant_cell_type group CXCL1 CXCL10 CXCL11
                     &lt;dbl&gt; &lt;dbl&gt;              &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1                       NA    NA                 NA    NA  4.72   8.52   4.51
# ℹ 4 more variables: CXCL12 &lt;dbl&gt;, CXCL13 &lt;dbl&gt;, CXCL14 &lt;dbl&gt;, CXCL2 &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Another useful example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span> </span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="at">.cols =</span> <span class="fu">everything</span>(), </span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.fns =</span> <span class="sc">~</span> <span class="fu">sum</span>(<span class="fu">is.na</span>(.))))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="using-where-to-select-numeric-columns" class="level3">
<h3 class="anchored" data-anchor-id="using-where-to-select-numeric-columns">Using <code>where()</code> to Select Numeric Columns</h3>
<p>We will probably not want to calculate means on non-numeric columns, so let’s select only numeric columns. For that we need another helper caller <code>where()</code> that lets us select columns based on their properties, like data type.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="at">.cols =</span> <span class="fu">where</span>(<span class="at">fn =</span> is.numeric), <span class="co"># fn is a function that returns TRUE or FALSE</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.fns =</span> mean)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="dv">5</span><span class="sc">:</span><span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 11
  percent_tumor_cells percent_not_cancer_cells CXCL1 CXCL10 CXCL11 CXCL12 CXCL13
                &lt;dbl&gt;                    &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1                  NA                       NA  4.72   8.52   4.51   6.72   4.35
# ℹ 4 more variables: CXCL14 &lt;dbl&gt;, CXCL2 &lt;dbl&gt;, CXCL3 &lt;dbl&gt;, CXCL5 &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
<section id="other-column-selection-helpers" class="level3">
<h3 class="anchored" data-anchor-id="other-column-selection-helpers">Other Column Selection Helpers</h3>
<p>There is another group of helpers that refers to column names:</p>
<ul>
<li><code>starts_with()</code></li>
<li><code>ends_with()</code></li>
<li><code>contains()</code></li>
</ul>
<p>And we can use them to select only columns starting with ‘CXC’:</p>
<p>All these ways can be used to select columns in tidyverse, also outside of <code>across()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Columns that start with "CXC"</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="fu">starts_with</span>(<span class="st">'CXC'</span>), mean))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 11
  CXCL1 CXCL10 CXCL11 CXCL12 CXCL13 CXCL14 CXCL2 CXCL3 CXCL5 CXCL6 CXCL9
  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1  4.72   8.52   4.51   6.72   4.35   4.41  5.88  3.82  3.43  3.46  5.90</code></pre>
</div>
</div>
<p>Lastly, we can always straight up supply the names of the columns we want to select in a vector:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Specific columns listed manually</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="fu">c</span>(CXCL3, CXCL5), mean))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 2
  CXCL3 CXCL5
  &lt;dbl&gt; &lt;dbl&gt;
1  3.82  3.43</code></pre>
</div>
</div>
<p>These helpers make your code cleaner, more scalable, and easier to maintain.</p>
</section>
<section id="summarise-becomes-more-powerful" class="level3">
<h3 class="anchored" data-anchor-id="summarise-becomes-more-powerful"><code>summarise()</code> becomes more powerful!</h3>
<p>So far, we’ve only applied a single function. But why stop at the mean? What if you want <strong>multiple statistics</strong> like mean, SD, min, and max — all in one go?</p>
<p>With <code>across()</code>, you can pass a <strong>list of functions</strong>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="at">.cols =</span> <span class="fu">starts_with</span>(<span class="st">"CXC"</span>), </span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.fns =</span> <span class="fu">list</span>(mean, sd, min, max)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 44
  CXCL1_1 CXCL1_2 CXCL1_3 CXCL1_4 CXCL10_1 CXCL10_2 CXCL10_3 CXCL10_4 CXCL11_1
    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;
1    4.72    1.62    2.43    10.3     8.52     2.00     3.26     12.1     4.51
# ℹ 35 more variables: CXCL11_2 &lt;dbl&gt;, CXCL11_3 &lt;dbl&gt;, CXCL11_4 &lt;dbl&gt;,
#   CXCL12_1 &lt;dbl&gt;, CXCL12_2 &lt;dbl&gt;, CXCL12_3 &lt;dbl&gt;, CXCL12_4 &lt;dbl&gt;,
#   CXCL13_1 &lt;dbl&gt;, CXCL13_2 &lt;dbl&gt;, CXCL13_3 &lt;dbl&gt;, CXCL13_4 &lt;dbl&gt;,
#   CXCL14_1 &lt;dbl&gt;, CXCL14_2 &lt;dbl&gt;, CXCL14_3 &lt;dbl&gt;, CXCL14_4 &lt;dbl&gt;,
#   CXCL2_1 &lt;dbl&gt;, CXCL2_2 &lt;dbl&gt;, CXCL2_3 &lt;dbl&gt;, CXCL2_4 &lt;dbl&gt;, CXCL3_1 &lt;dbl&gt;,
#   CXCL3_2 &lt;dbl&gt;, CXCL3_3 &lt;dbl&gt;, CXCL3_4 &lt;dbl&gt;, CXCL5_1 &lt;dbl&gt;, CXCL5_2 &lt;dbl&gt;,
#   CXCL5_3 &lt;dbl&gt;, CXCL5_4 &lt;dbl&gt;, CXCL6_1 &lt;dbl&gt;, CXCL6_2 &lt;dbl&gt;, …</code></pre>
</div>
</div>
<p>This gives you one wide row per column, with new columns like <code>CXCL1_1</code>, <code>CXCL1_2</code>, etc. A bit cryptic, right?</p>
<p>Let’s clean it up by naming the functions and columns:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>gene_summary <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="at">.cols =</span> <span class="fu">starts_with</span>(<span class="st">"CXC"</span>), </span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.fns =</span> <span class="fu">list</span>(<span class="at">mean =</span> mean, </span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>                               <span class="at">sd =</span> sd, </span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>                               <span class="at">min =</span> min, </span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>                               <span class="at">max =</span> max),</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.names =</span> <span class="st">"{.col}-{.fn}"</span>))</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>gene_summary</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 44
  `CXCL1-mean` `CXCL1-sd` `CXCL1-min` `CXCL1-max` `CXCL10-mean` `CXCL10-sd`
         &lt;dbl&gt;      &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;         &lt;dbl&gt;       &lt;dbl&gt;
1         4.72       1.62        2.43        10.3          8.52        2.00
# ℹ 38 more variables: `CXCL10-min` &lt;dbl&gt;, `CXCL10-max` &lt;dbl&gt;,
#   `CXCL11-mean` &lt;dbl&gt;, `CXCL11-sd` &lt;dbl&gt;, `CXCL11-min` &lt;dbl&gt;,
#   `CXCL11-max` &lt;dbl&gt;, `CXCL12-mean` &lt;dbl&gt;, `CXCL12-sd` &lt;dbl&gt;,
#   `CXCL12-min` &lt;dbl&gt;, `CXCL12-max` &lt;dbl&gt;, `CXCL13-mean` &lt;dbl&gt;,
#   `CXCL13-sd` &lt;dbl&gt;, `CXCL13-min` &lt;dbl&gt;, `CXCL13-max` &lt;dbl&gt;,
#   `CXCL14-mean` &lt;dbl&gt;, `CXCL14-sd` &lt;dbl&gt;, `CXCL14-min` &lt;dbl&gt;,
#   `CXCL14-max` &lt;dbl&gt;, `CXCL2-mean` &lt;dbl&gt;, `CXCL2-sd` &lt;dbl&gt;, …</code></pre>
</div>
</div>
<p>Much better! Now the column names are readable and include both the variable and the statistic.</p>
<p>Still not your preferred format?</p>
<p>You can probably pivot your way out of that!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>gene_summary <span class="sc">%&gt;%</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">everything</span>(), </span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">"gene"</span>, <span class="st">"statistic"</span>), </span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_sep =</span> <span class="st">"-"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> statistic, </span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">values_from =</span> value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 11 × 5
   gene    mean    sd   min   max
   &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 CXCL1   4.72 1.62   2.43 10.3 
 2 CXCL10  8.52 2.00   3.26 12.1 
 3 CXCL11  4.51 1.59   2.49  9.48
 4 CXCL12  6.72 1.59   3.64 11.1 
 5 CXCL13  4.35 1.20   3.01  9.57
 6 CXCL14  4.41 1.82   2.16  9.99
 7 CXCL2   5.88 1.70   3.06 11.8 
 8 CXCL3   3.82 0.659  2.70  7.78
 9 CXCL5   3.43 1.11   2.38 10.7 
10 CXCL6   3.46 1.01   2.48 10.3 
11 CXCL9   5.90 1.92   2.90 11.4 </code></pre>
</div>
</div>
<p>Now you get a long format table with one row per variable and all your stats in columns — clean, tidy, and ready for interpretation.</p>
<p><img src="../figures/satisfying_meme.jpg" class="img-fluid" width="200"></p>
</section>
<section id="the-anonymous-function-and-." class="level3">
<h3 class="anchored" data-anchor-id="the-anonymous-function-and-.">The anonymous function: <code>~</code> and <code>.</code></h3>
<p>We promised to get back to handling the <code>NA</code>s when doing summary stats. Let’s just add the <code>na.rm=TRUE</code> argument. To not have too many things going on at once we’ll only do <code>mean()</code> for now:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="at">.cols =</span> <span class="fu">where</span>(is.numeric), </span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.fns =</span> <span class="fu">list</span>(<span class="at">mean =</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)),</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.names =</span> <span class="st">"{.col}-{.fn}"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in `summarise()`:
ℹ In argument: `across(...)`.
Caused by error in `mean.default()`:
! argument "x" is missing, with no default</code></pre>
</div>
</div>
<p>Brrrtt! We may not.</p>
<p>Why doesn’t this work?</p>
<p>When you pass functions directly into <code>across()</code> using the <strong>shorthand syntax</strong> (<code>mean</code>, <code>sd</code>, etc.), you’re only allowed to use the <strong>bare function with default arguments</strong>. You will also notice that we didn’t use brackets after their names, which is part of using the function shorthanded. Once you try to add something like <code>na.rm = TRUE</code>, the shorthandness breaks.</p>
<p>To pass arguments to a function that is called inside another function (i.e.&nbsp;calling <code>mean</code> inside <code>summarise</code>), we need to use what’s called an <strong>anonymous function</strong>. Don’t worry — it’s not as scary as it sounds.</p>
<p>It is written as a <code>~</code> and looks like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="at">.cols =</span> <span class="fu">where</span>(is.numeric), </span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.fns =</span> <span class="fu">list</span>(<span class="at">mean =</span> <span class="sc">~</span> <span class="fu">mean</span>(., <span class="at">na.rm =</span> <span class="cn">TRUE</span>)),</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.names =</span> <span class="st">"{.col}-{.fn}"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 17
  `age_at_initial_path_diagn-mean` `days_to_death-mean` percent_normal_cells-m…¹
                             &lt;dbl&gt;                &lt;dbl&gt;                    &lt;dbl&gt;
1                             59.7                1010.                     2.38
# ℹ abbreviated name: ¹​`percent_normal_cells-mean`
# ℹ 14 more variables: `percent_stromal_cells-mean` &lt;dbl&gt;,
#   `percent_tumor_cells-mean` &lt;dbl&gt;, `percent_not_cancer_cells-mean` &lt;dbl&gt;,
#   `CXCL1-mean` &lt;dbl&gt;, `CXCL10-mean` &lt;dbl&gt;, `CXCL11-mean` &lt;dbl&gt;,
#   `CXCL12-mean` &lt;dbl&gt;, `CXCL13-mean` &lt;dbl&gt;, `CXCL14-mean` &lt;dbl&gt;,
#   `CXCL2-mean` &lt;dbl&gt;, `CXCL3-mean` &lt;dbl&gt;, `CXCL5-mean` &lt;dbl&gt;,
#   `CXCL6-mean` &lt;dbl&gt;, `CXCL9-mean` &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Let’s break it down:</p>
<ul>
<li><p><code>~</code> to define the function. <code>~ mean(., na.rm = TRUE)</code> tells R:<br>
<em>“for each column, compute the mean, ignoring NAs”</em></p></li>
<li><p><code>.</code> is a placeholder for the current column being operated or ‘the data previously referred to’. We need to use the <code>.</code> because <code>mean</code> when called as a proper function needs to have an argument (a vector of numbers) to work on.</p></li>
</ul>
<p>Compare it to this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(df_comb<span class="sc">$</span>age_at_initial_path_diagn, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 59.69542</code></pre>
</div>
</div>
<p>In both cases, the column must be explicitly passed in and <code>.</code> serves that role inside anonymous functions.</p>
</section>
<section id="multiple-summary-statistics-with-na.rm-true" class="level3">
<h3 class="anchored" data-anchor-id="multiple-summary-statistics-with-na.rm-true">Multiple Summary Statistics with <code>na.rm = TRUE</code></h3>
<p>Now let’s compute <strong>several</strong> stats per column, all with <code>na.rm = TRUE</code>. Here we chose just the integer columns:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>stats <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="at">.cols =</span> <span class="fu">where</span>(is.integer),</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.fns =</span> <span class="fu">list</span>(<span class="at">mean =</span> <span class="sc">~</span> <span class="fu">mean</span>(., <span class="at">na.rm =</span> <span class="cn">TRUE</span>),  </span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>                               <span class="at">sd =</span> <span class="sc">~</span> <span class="fu">sd</span>(., <span class="at">na.rm =</span> <span class="cn">TRUE</span>), </span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>                               <span class="at">min =</span> <span class="sc">~</span> <span class="fu">min</span>(., <span class="at">na.rm =</span> <span class="cn">TRUE</span>), </span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>                               <span class="at">max =</span> <span class="sc">~</span> <span class="fu">max</span>(., <span class="at">na.rm =</span> <span class="cn">TRUE</span>)),</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.names =</span> <span class="st">"{.col}-{.fn}"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add reformating</span></span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">everything</span>(), </span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">"variable"</span>, <span class="st">"statistic"</span>), </span>
<span id="cb57-11"><a href="#cb57-11" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_sep =</span> <span class="st">"-"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb57-12"><a href="#cb57-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> statistic, </span>
<span id="cb57-13"><a href="#cb57-13" aria-hidden="true" tabindex="-1"></a>              <span class="at">values_from =</span> value) </span>
<span id="cb57-14"><a href="#cb57-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-15"><a href="#cb57-15" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(stats)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  variable                     mean     sd   min   max
  &lt;chr&gt;                       &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 age_at_initial_path_diagn   59.7   11.6     26    89
2 days_to_death             1010.   803.       8  5480
3 percent_normal_cells         2.38   6.68     0    55
4 percent_stromal_cells       12.8   11.9      0    70
5 percent_tumor_cells         80.6   15.4      0   100
6 percent_not_cancer_cells    15.1   12.3      0    97</code></pre>
</div>
</div>
<p>This produces a tidy table with one row per gene and columns for <code>mean</code>, <code>sd</code>, <code>min</code>, and <code>max</code>.</p>
<p>The helper functions can be used in other tidyverse operation such as <code>mutate</code> and <code>select</code>.</p>
</section>
</section>
<section id="handling-outliers" class="level2">
<h2 class="anchored" data-anchor-id="handling-outliers">Handling Outliers</h2>
<p>Now that we’ve explored summary statistics, it’s time to consider <strong>outliers</strong> — data points that differ markedly from the rest. Outliers can be caused by measurement or entry errors, natural variability or rare events or anomalies.</p>
<p>Outliers can skew your summary statistics, distort visualizations and can affect model performance, especially in regression.</p>
<p>Let’s use our tidyverse skills to <strong>calculate some basic stats</strong> and <strong>visualize potential outliers</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>stats <span class="ot">&lt;-</span> stats <span class="sc">%&gt;%</span> </span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">threshold_upper =</span> mean <span class="sc">+</span> <span class="dv">2</span> <span class="sc">*</span> sd,</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">threshold_lower =</span> mean <span class="sc">-</span> <span class="dv">2</span> <span class="sc">*</span> sd) <span class="sc">%&gt;%</span></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(threshold_lower, threshold_upper), <span class="at">names_to =</span> <span class="st">"threshold_type"</span>, <span class="at">values_to =</span> <span class="st">"threshold"</span>)</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(stats)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12 × 7
   variable                     mean     sd   min   max threshold_type threshold
   &lt;chr&gt;                       &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;              &lt;dbl&gt;
 1 age_at_initial_path_diagn   59.7   11.6     26    89 threshold_low…     36.6 
 2 age_at_initial_path_diagn   59.7   11.6     26    89 threshold_upp…     82.8 
 3 days_to_death             1010.   803.       8  5480 threshold_low…   -596.  
 4 days_to_death             1010.   803.       8  5480 threshold_upp…   2615.  
 5 percent_normal_cells         2.38   6.68     0    55 threshold_low…    -11.0 
 6 percent_normal_cells         2.38   6.68     0    55 threshold_upp…     15.8 
 7 percent_stromal_cells       12.8   11.9      0    70 threshold_low…    -10.9 
 8 percent_stromal_cells       12.8   11.9      0    70 threshold_upp…     36.6 
 9 percent_tumor_cells         80.6   15.4      0   100 threshold_low…     49.8 
10 percent_tumor_cells         80.6   15.4      0   100 threshold_upp…    111.  
11 percent_not_cancer_cells    15.1   12.3      0    97 threshold_low…     -9.42
12 percent_not_cancer_cells    15.1   12.3      0    97 threshold_upp…     39.6 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>df_comb_longer <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span> </span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">where</span>(is.integer),</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"variable"</span>,</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"value"</span>)</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(df_comb_longer)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3,468 × 25
   unique_patient_ID sample_recode summarygrade summarystage tumorstage grade
   &lt;chr&gt;             &lt;chr&gt;         &lt;fct&gt;        &lt;fct&gt;        &lt;fct&gt;      &lt;fct&gt;
 1 TCGA-20-0987      Tumor         high         late         3          3    
 2 TCGA-20-0987      Tumor         high         late         3          3    
 3 TCGA-20-0987      Tumor         high         late         3          3    
 4 TCGA-20-0987      Tumor         high         late         3          3    
 5 TCGA-20-0987      Tumor         high         late         3          3    
 6 TCGA-20-0987      Tumor         high         late         3          3    
 7 TCGA-23-1031      Tumor         high         late         4          3    
 8 TCGA-23-1031      Tumor         high         late         4          3    
 9 TCGA-23-1031      Tumor         high         late         4          3    
10 TCGA-23-1031      Tumor         high         late         4          3    
# ℹ 3,458 more rows
# ℹ 19 more variables: recurrence_status &lt;fct&gt;, vital_status &lt;fct&gt;,
#   batch &lt;fct&gt;, stage &lt;chr&gt;, dominant_cell_type &lt;chr&gt;, group &lt;chr&gt;,
#   CXCL1 &lt;dbl&gt;, CXCL10 &lt;dbl&gt;, CXCL11 &lt;dbl&gt;, CXCL12 &lt;dbl&gt;, CXCL13 &lt;dbl&gt;,
#   CXCL14 &lt;dbl&gt;, CXCL2 &lt;dbl&gt;, CXCL3 &lt;dbl&gt;, CXCL5 &lt;dbl&gt;, CXCL6 &lt;dbl&gt;,
#   CXCL9 &lt;dbl&gt;, variable &lt;chr&gt;, value &lt;int&gt;</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_comb_longer, </span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> value)) <span class="sc">+</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">data =</span> stats, </span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>             <span class="fu">aes</span>(<span class="at">xintercept =</span> threshold),</span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(variable), <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span> </span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 128 rows containing non-finite outside the scale range
(`stat_bin()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Sometimes, visualizing values in relation to <strong>other variables</strong> can make potential outliers easier to spot. Below, you’ll notice that a patient with 0% cancer cells and 0% tumor cells likely represents an outlier:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Bivariate scatter plot colored by stromal cell percentage</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_comb, </span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> percent_tumor_cells, <span class="at">y =</span> percent_not_cancer_cells, <span class="at">color =</span> percent_stromal_cells)) <span class="sc">+</span></span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>() <span class="sc">+</span></span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>There are many different tools and strategies available to detect and handle outliers. However, in this case, we’ll keep it simple: we will remove patients where both <code>percent_tumor_cells</code> and <code>percent_not_cancer_cells</code> are less than 5%.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">isOutlier =</span> (percent_tumor_cells <span class="sc">&lt;</span> <span class="dv">5</span> <span class="sc">&amp;</span> percent_not_cancer_cells <span class="sc">&lt;</span> <span class="dv">5</span>))</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(isOutlier <span class="sc">==</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 7 × 30
  unique_patient_ID sample_recode summarygrade summarystage tumorstage grade
  &lt;chr&gt;             &lt;chr&gt;         &lt;fct&gt;        &lt;fct&gt;        &lt;fct&gt;      &lt;fct&gt;
1 TCGA-24-1103      Tumor         high         late         3          3    
2 TCGA-23-1119      Tumor         high         late         3          3    
3 TCGA-23-1107      Tumor         high         late         4          3    
4 TCGA-23-1120      Tumor         high         late         3          3    
5 TCGA-23-1121      Tumor         high         late         3          3    
6 TCGA-23-1123      Tumor         high         late         3          3    
7 TCGA-23-1124      Tumor         high         late         3          3    
# ℹ 24 more variables: age_at_initial_path_diagn &lt;int&gt;,
#   recurrence_status &lt;fct&gt;, days_to_death &lt;int&gt;, vital_status &lt;fct&gt;,
#   percent_normal_cells &lt;int&gt;, percent_stromal_cells &lt;int&gt;,
#   percent_tumor_cells &lt;int&gt;, batch &lt;fct&gt;, percent_not_cancer_cells &lt;int&gt;,
#   stage &lt;chr&gt;, dominant_cell_type &lt;chr&gt;, group &lt;chr&gt;, CXCL1 &lt;dbl&gt;,
#   CXCL10 &lt;dbl&gt;, CXCL11 &lt;dbl&gt;, CXCL12 &lt;dbl&gt;, CXCL13 &lt;dbl&gt;, CXCL14 &lt;dbl&gt;,
#   CXCL2 &lt;dbl&gt;, CXCL3 &lt;dbl&gt;, CXCL5 &lt;dbl&gt;, CXCL6 &lt;dbl&gt;, CXCL9 &lt;dbl&gt;, …</code></pre>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span> </span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(isOutlier <span class="sc">==</span> <span class="cn">FALSE</span> <span class="sc">|</span> <span class="fu">is.na</span>(isOutlier))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p><strong>Be cautious when removing data</strong>: Outlier removal should always be guided by domain knowledge and clear justification. Removing too much — or for the wrong reasons — can distort your analysis.</p>
</blockquote>
</section>
<section id="sample-outliers-across-variables" class="level2">
<h2 class="anchored" data-anchor-id="sample-outliers-across-variables">Sample outliers across variables</h2>
<p>Instead of visualizing variables one by one, we can also look at samples across a set of variables and identify outliers that way. If we have a lot of integer and numeric values in our dataset, one thing that might make sense is to make a dendogram (tree) using hierarchical clustering and identify if any of the samples ‘stand out’.</p>
<p>Let’s try this for the integer variables in our dataset.</p>
<p>We first select all integer columns, drop missing values (they are non-informative for clustering anyways). Then, we make a plot to get an idea of data distributions:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>df_int <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">where</span>(is.integer)) <span class="sc">%&gt;%</span></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>()</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>df_int <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 6
  age_at_initial_path_diagn days_to_death percent_normal_cells
                      &lt;int&gt;         &lt;int&gt;                &lt;int&gt;
1                        61           701                   16
2                        60           574                    8
3                        53          1264                    0
4                        78            61                    0
5                        74           789                   12
6                        73            84                   15
# ℹ 3 more variables: percent_stromal_cells &lt;int&gt;, percent_tumor_cells &lt;int&gt;,
#   percent_not_cancer_cells &lt;int&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>df_int <span class="sc">%&gt;%</span> </span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">where</span>(is.integer),</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"variable"</span>,</span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"value"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(., <span class="fu">aes</span>(<span class="at">x =</span> value)) <span class="sc">+</span></span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">color=</span> <span class="st">'white'</span>) <span class="sc">+</span></span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(variable), <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span> </span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Next, we scale our data (more on this in Presentation 5.) and calculate the distances between all pairs of scaled variables and perform hierarchical clustering.</p>
<p>Do you not see the plot clearly? Copy-paste the code into the console and the plot will appear in the <em>Plots</em> window. Press <em>Export</em> –&gt; <em>Copy to Clipboard…</em> and drag out the plot as large as you can. Alternatively, save it in a large format and look at it outside of R studio.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Euclidean pairwise distances</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>df_int_dist <span class="ot">&lt;-</span> df_int <span class="sc">%&gt;%</span> </span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">everything</span>(), scale)) <span class="sc">%&gt;%</span></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist</span>(., <span class="at">method =</span> <span class="st">'euclidean'</span>)</span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Hierarchical clustering with Ward's distance metric</span></span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a>hclust_int <span class="ot">&lt;-</span> <span class="fu">hclust</span>(df_int_dist, <span class="at">method =</span> <span class="st">'ward.D2'</span>)</span>
<span id="cb73-8"><a href="#cb73-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-9"><a href="#cb73-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the result of clustering</span></span>
<span id="cb73-10"><a href="#cb73-10" aria-hidden="true" tabindex="-1"></a><span class="co"># png("dendrogram_plot.png", width = 2000, height = 500)</span></span>
<span id="cb73-11"><a href="#cb73-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hclust_int, <span class="at">main =</span> <span class="st">"Clustering based on scaled integer values"</span>, <span class="at">cex =</span> <span class="fl">0.7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="presentation2_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="co"># dev.off()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>From our plot we see that a couple of samples seem to have very short branch lengths (which may indicate an outlier). One such sample is 208 (row id). Lets have a look at it:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>df_int <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">where</span>(is.integer)) <span class="sc">%&gt;%</span></span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>() </span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a>df_int[<span class="dv">208</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 6
  age_at_initial_path_diagn days_to_death percent_normal_cells
                      &lt;int&gt;         &lt;int&gt;                &lt;int&gt;
1                        74           197                   55
# ℹ 3 more variables: percent_stromal_cells &lt;int&gt;, percent_tumor_cells &lt;int&gt;,
#   percent_not_cancer_cells &lt;int&gt;</code></pre>
</div>
</div>
<p>We can now compare the values of this sample to the distributions above (histograms) to try an figure out what makes it an outlier and IF it should be removed. For the sample in row 208, age at diagnosis and days to death is with in normal range, though percent stromal and not tumor cell is in the very high end while percent tumor cells is in the very low end. These are pretty extreme values for several of the variables - why one would consider removing this sample.</p>
</section>
<section id="handling-missing-data" class="level2">
<h2 class="anchored" data-anchor-id="handling-missing-data">Handling Missing Data</h2>
<p>Missing data is common in most real-world datasets and can significantly affect the quality of our analysis. During EDA, it’s essential to <strong>identify</strong>, <strong>understand</strong>, and <strong>properly handle</strong> missing values to avoid biased or misleading conclusions.</p>
<p>Missing data can fall into three categories:</p>
<ul>
<li><p><strong>Missing Completely At Random (MCAR)</strong>: The missingness is independent of both the observed and unobserved data.</p></li>
<li><p><strong>Missing At Random (MAR)</strong>: The missingness is related to the observed data.</p></li>
<li><p><strong>Missing Not At Random (MNAR)</strong>: The missingness depends on the observed values themselves.</p></li>
</ul>
<p>Understanding which category your data falls into helps guide how to deal with it. Let’s inspect rows with the most missing values:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="sc">%&gt;%</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">na_count =</span> <span class="fu">rowSums</span>(<span class="fu">is.na</span>(.))) <span class="sc">%&gt;%</span></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(na_count)) <span class="sc">%&gt;%</span></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_head</span>(<span class="at">n =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 31
   unique_patient_ID sample_recode summarygrade summarystage tumorstage grade
   &lt;chr&gt;             &lt;chr&gt;         &lt;fct&gt;        &lt;fct&gt;        &lt;fct&gt;      &lt;fct&gt;
 1 TCGA-01-0630      Healthy       &lt;NA&gt;         &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt; 
 2 TCGA-01-0631      Healthy       &lt;NA&gt;         &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt; 
 3 TCGA-01-0633      Healthy       &lt;NA&gt;         &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt; 
 4 TCGA-01-0636      Healthy       &lt;NA&gt;         &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt; 
 5 TCGA-01-0637      Healthy       &lt;NA&gt;         &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt; 
 6 TCGA-01-0628      Healthy       &lt;NA&gt;         &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt; 
 7 TCGA-01-0639      Healthy       &lt;NA&gt;         &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt; 
 8 TCGA-01-0642      Healthy       &lt;NA&gt;         &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt; 
 9 TCGA-23-1023      &lt;NA&gt;          high         late         3          3    
10 TCGA-04-1353      Tumor         &lt;NA&gt;         &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt; 
# ℹ 25 more variables: age_at_initial_path_diagn &lt;int&gt;,
#   recurrence_status &lt;fct&gt;, days_to_death &lt;int&gt;, vital_status &lt;fct&gt;,
#   percent_normal_cells &lt;int&gt;, percent_stromal_cells &lt;int&gt;,
#   percent_tumor_cells &lt;int&gt;, batch &lt;fct&gt;, percent_not_cancer_cells &lt;int&gt;,
#   stage &lt;chr&gt;, dominant_cell_type &lt;chr&gt;, group &lt;chr&gt;, CXCL1 &lt;dbl&gt;,
#   CXCL10 &lt;dbl&gt;, CXCL11 &lt;dbl&gt;, CXCL12 &lt;dbl&gt;, CXCL13 &lt;dbl&gt;, CXCL14 &lt;dbl&gt;,
#   CXCL2 &lt;dbl&gt;, CXCL3 &lt;dbl&gt;, CXCL5 &lt;dbl&gt;, CXCL6 &lt;dbl&gt;, CXCL9 &lt;dbl&gt;, …</code></pre>
</div>
</div>
<p>Here we notice that most of the missing values come from samples where the patients were (<strong>healthy (sample_recode)</strong> - this points to either MAR or MNAR. Unless however we have evidence that, missingness is directly related to the observed values themselves, we prefer to assume MAR.</p>
<p>There are two main approaches when deciding what to do:</p>
<ul>
<li><p><strong>Remove rows</strong>&nbsp;with missing values — useful when the proportion is small and missingness is MCAR.</p></li>
<li><p><strong>Impute values</strong>&nbsp;— helps preserve sample size and may improve robustness, especially when missingness is MAR.</p></li>
</ul>
<p>Consider the impact of missing data. Even after imputing, missing data can cause uncertainty and bias so understands the result with caution.</p>
<blockquote class="blockquote">
<p><strong>Caution</strong>: Even after imputation, results may be biased. Always interpret findings with an understanding of how missing data was handled.</p>
</blockquote>
<p>Since we have very few samples with a high number of missing values AND we are not interested in working with these likely either misclassified or mislabled samples in our analysis, we can safely them:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>df_comb <span class="ot">&lt;-</span> df_comb <span class="sc">%&gt;%</span> <span class="fu">filter</span>(sample_recode <span class="sc">==</span> <span class="st">"Tumor"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We’ll keep the remaining&nbsp;<code>NA</code>s as-is for now, depending on context.</p>
<p><br> <br></p>
</section>
<section id="in-summary" class="level2">
<h2 class="anchored" data-anchor-id="in-summary">In Summary</h2>
<p>Key Takeaways</p>
<ul>
<li><p>Long format is your&nbsp;<strong>best friend</strong>&nbsp;when plotting or modeling.</p></li>
<li><p>Use&nbsp;<code>across()</code>&nbsp;and helpers to write&nbsp;<strong>clean, scalable summaries</strong>.</p></li>
<li><p>Visual checks are as important as statistical checks.</p></li>
<li><p><strong>EDA is not optional</strong>&nbsp;— it’s the foundation of trustworthy analysis.</p></li>
<li><p>Keep your workflow&nbsp;<strong>reproducible</strong>&nbsp;and&nbsp;<strong>tidy</strong>&nbsp;for future modeling stages.</p></li>
</ul>
<p>We have conducted an initial&nbsp;<strong>exploratory data analysis (EDA)</strong>&nbsp;and overall, the dataset looks clean and well-structured.</p>
<ul>
<li><p>We identified and&nbsp;<strong>removed a few samples</strong>, as they had many missing values.</p></li>
<li><p>One sample was also removed due to inconsistencies in cell type proportions.</p></li>
<li><p>We observed that the categories within stage and grade are&nbsp;<strong>not fully balanced</strong>, therefore we decided to use the&nbsp;<strong>summary stage</strong>&nbsp;and&nbsp;<strong>summary grade</strong>&nbsp;variables</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(df_comb, <span class="at">file=</span><span class="st">"../data/Ovarian_comb_prep.RData"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>